diff -N -r --unified glibc-2.25.orig/csu/libc-start.c glibc-2.25/csu/libc-start.c
--- glibc-2.25.orig/csu/libc-start.c	2017-09-11 17:14:55.241208737 +0000
+++ glibc-2.25/csu/libc-start.c	2017-09-12 06:55:39.618444139 +0000
@@ -104,6 +104,234 @@
 # define MAIN_AUXVEC_PARAM
 #endif
 
+#ifdef USE_MVEE_LIBC
+
+#define MVEE_SLAVE_YIELD
+#define MVEE_TOTAL_CLOCK_COUNT   2048
+#define MVEE_CLOCK_GROUP_SIZE    64
+#define MVEE_TOTAL_CLOCK_GROUPS  (MVEE_TOTAL_CLOCK_COUNT / MVEE_CLOCK_GROUP_SIZE)
+
+struct mvee_counter
+{
+  volatile unsigned long lock;
+  volatile unsigned long counter;
+  unsigned char padding[64 - 2 * sizeof(unsigned long)]; // prevents false sharing
+};
+
+struct mvee_op_entry
+{
+  volatile unsigned long  counter_and_idx; // the value we must see in mvee_counters[idx] before we can replay the operation
+};
+
+static unsigned char                  mvee_sync_enabled             = 0;
+static unsigned char                  mvee_libc_initialized         = 0;
+static unsigned char                  mvee_master_variant           = 0;
+static __thread unsigned long         mvee_thread_local_pos         = 0; // our position in the thread local queue
+static __thread  
+    struct mvee_op_entry*             mvee_thread_local_queue       = NULL;
+static __thread unsigned long         mvee_thread_local_queue_size  = 0; // nr of slots in the thread local queue
+static __thread unsigned short        mvee_prev_idx                 = 0;
+
+__attribute__((aligned (64)))
+static struct mvee_counter            mvee_counters[MVEE_TOTAL_CLOCK_COUNT + 1];
+
+#define likely(x)       __builtin_expect((x),1)
+#define unlikely(x)     __builtin_expect((x),0)
+#define gcc_barrier()  asm volatile("" ::: "memory")
+#ifndef arch_cpu_relax
+#define arch_cpu_relax()
+#endif
+
+void mvee_invalidate_buffer(void)
+{
+	mvee_thread_local_queue = NULL;
+}
+
+unsigned char mvee_atomic_preop_internal(volatile void* word_ptr)
+{
+	if (unlikely(!mvee_sync_enabled))
+		return 0;
+
+	if (unlikely(!mvee_thread_local_queue))
+    {
+		long mvee_thread_local_queue_id = syscall(MVEE_GET_SHARED_BUFFER, &mvee_counters, MVEE_LIBC_ATOMIC_BUFFER, &mvee_thread_local_queue_size, &mvee_thread_local_pos, NULL);
+		mvee_thread_local_queue_size   /= sizeof(struct mvee_op_entry);
+		mvee_thread_local_queue         = (void*)syscall(__NR_shmat, mvee_thread_local_queue_id, NULL, 0);     
+    }
+
+	if (unlikely(mvee_thread_local_pos >= mvee_thread_local_queue_size))
+    {
+		syscall(MVEE_FLUSH_SHARED_BUFFER, MVEE_LIBC_ATOMIC_BUFFER);
+		mvee_thread_local_pos = 0;
+    }
+
+	if (likely(mvee_master_variant))
+    {
+		// page number defines the clock group
+		// offset within page defines the clock within that group
+		mvee_prev_idx = (((((unsigned long)word_ptr >> 24) % MVEE_TOTAL_CLOCK_GROUPS) * (MVEE_CLOCK_GROUP_SIZE) 
+						  + ((((unsigned long)word_ptr & 4095) >> 6) % MVEE_CLOCK_GROUP_SIZE))
+						 & 0xFFF) + 1;
+
+		while (!__sync_bool_compare_and_swap(&mvee_counters[mvee_prev_idx].lock, 0, 1))
+			arch_cpu_relax();
+		
+		unsigned long pos = mvee_counters[mvee_prev_idx].counter;    
+
+		mvee_thread_local_queue[mvee_thread_local_pos++].counter_and_idx 
+			= (pos << 12) | mvee_prev_idx;
+
+		atomic_full_barrier();
+
+		return 1;
+    }
+	else
+    {
+		unsigned long counter_and_idx = 0;
+
+		while (unlikely(1))
+		{
+			counter_and_idx = mvee_thread_local_queue[mvee_thread_local_pos].counter_and_idx;
+
+			if (likely(counter_and_idx))
+				break;
+
+#ifdef MVEE_SLAVE_YIELD
+			syscall(__NR_sched_yield);
+#else
+			arch_cpu_relax();
+#endif
+		}
+
+		mvee_prev_idx = counter_and_idx & 0xFFF;
+		counter_and_idx &= ~0xFFF;
+
+		atomic_full_barrier();
+
+		while ((mvee_counters[mvee_prev_idx].counter << 12) != counter_and_idx)
+#ifdef MVEE_SLAVE_YIELD
+			syscall(__NR_sched_yield);
+#else
+		arch_cpu_relax();
+#endif
+
+		atomic_full_barrier();
+		
+		return 2;
+    }
+}
+
+void mvee_atomic_postop_internal(unsigned char preop_result)
+{
+	atomic_full_barrier();
+	
+	if(likely(preop_result) == 1)
+	{
+		gcc_barrier();
+		orig_atomic_increment(&mvee_counters[mvee_prev_idx].counter);
+		atomic_full_barrier();
+		mvee_counters[mvee_prev_idx].lock = 0;
+	}
+	else if (likely(preop_result) == 2)
+	{
+		gcc_barrier();
+		mvee_counters[mvee_prev_idx].counter++;
+		mvee_thread_local_pos++;
+	}
+}
+
+/* Checks if all variants got ALIGNMENT aligned heaps from
+   the previous mmap request. If some of them have not, ALL variants
+   have to bail out and fall back to another heap allocation method.
+   This ensures that the variants stay in sync with respect to future mm
+   requests.
+*/
+#define ALIGNMENT 0x4000000
+
+int
+mvee_all_heaps_aligned(char* heap, unsigned long alloc_size)
+{
+	// if we're not running under MVEE control,
+	// just check the alignment of the current heap
+	if (!mvee_thread_local_queue)
+	{
+		if ((unsigned long)heap & (ALIGNMENT-1))
+			return 0;
+		return 1;
+	}
+
+	// We ARE running under MVEE control
+	// => ask the MVEE to check the alignments
+	// of ALL heaps
+	return syscall(MVEE_ALL_HEAPS_ALIGNED, heap, ALIGNMENT, alloc_size);
+}
+
+/* 
+ * mvee_infinite_loop:
+ * this function is used for both thread transfering and signal delivery 
+ * 
+ * 1) to transfer threads to a new monitor, the original monitor (i.e. the 
+ * monitor that monitors the childs that instigated the fork event) needs to
+ * detach from the threads first. While the threads are detached, they can
+ * run freely, without the intervention of a debugger.
+ * As such, we have to move the program counter to an infinite loop while
+ * the threads are detached. This way, the threads will all be in an equivalent
+ * state when the new monitor attaches to them.
+ * Because we're going to replace the registers by their original contents
+ * when the new monitor attaches, we can use sys_pause calls in the infinite
+ * loop.
+ * 
+ * 2) delivering signals through the ptrace API happens asynchronously 
+ * (I found out the hard way). As such, we should wait for the threads to be
+ * in equivalent states (e.g. stopped on the same syscall). Then the registers
+ * should be backed up and the syscall nr should be replaced by a harmless
+ * syscall that doesn't modify the program state. We use sys_getpid for this
+ * purpose. When that replaced syscall returns, we change the pc to this
+ * infinite loop while we wait for async signal delivery.
+ * We probably cannot use syscalls while waiting for signal delivery. 
+ * One possible exception is sys_sched_yield. Our modified MVEE kernel does
+ * not report this syscall to the ptracer
+ * 
+ * the with_syscalls parameter is passed through the ecx register!
+ */
+void mvee_infinite_loop(int with_syscalls, int dummy)
+{
+	if (with_syscalls)
+    {
+		for (;;)
+			syscall(__NR_pause);
+    }
+	else
+    {
+		for (;;)
+		{
+			dummy = dummy << 2;
+		}
+    }
+}
+
+int mvee_should_sync_tid(void)
+{
+	return mvee_sync_enabled ? 1 : 0;
+}
+
+unsigned char mvee_atomic_preop(unsigned short op_type, void* word_ptr)
+{
+	return mvee_atomic_preop_internal(word_ptr);
+}
+
+void mvee_atomic_postop(unsigned char preop_result)
+{
+	mvee_atomic_postop_internal(preop_result);
+}
+
+unsigned char mvee_should_futex_unlock(void)
+{
+	return (!mvee_master_variant && mvee_sync_enabled) ? 1 : 0;
+}
+
+#endif
+
 STATIC int LIBC_START_MAIN (int (*main) (int, char **, char **
 					 MAIN_AUXVEC_DECL),
 			    int argc,
@@ -274,6 +502,11 @@
 #ifndef SHARED
   _dl_debug_initialize (0, LM_ID_BASE);
 #endif
+
+  (void) syscall(MVEE_RUNS_UNDER_MVEE_CONTROL, &mvee_sync_enabled, &mvee_infinite_loop, 
+				 NULL, NULL, &mvee_master_variant);
+  mvee_libc_initialized = 1;
+
 #ifdef HAVE_CLEANUP_JMP_BUF
   /* Memory for the cancellation buffer.  */
   struct pthread_unwind_buf unwind_buf;
diff -N -r --unified glibc-2.25.orig/csu/Versions glibc-2.25/csu/Versions
--- glibc-2.25.orig/csu/Versions	2017-09-11 17:14:55.239208594 +0000
+++ glibc-2.25/csu/Versions	2017-09-11 22:18:57.814846194 +0000
@@ -2,6 +2,14 @@
   GLIBC_2.0 {
     # helper functions
     __libc_init_first; __libc_start_main;
+    mvee_atomic_preop;
+    mvee_atomic_postop;
+    mvee_atomic_preop_internal;
+    mvee_atomic_postop_internal;
+    mvee_invalidate_buffer;
+    mvee_all_heaps_aligned;
+    mvee_should_sync_tid;
+    mvee_should_futex_unlock;
   }
   GLIBC_2.1 {
     # New special glibc functions.
diff -N -r --unified glibc-2.25.orig/elf/rtld.c glibc-2.25/elf/rtld.c
--- glibc-2.25.orig/elf/rtld.c	2017-09-11 17:14:55.334215390 +0000
+++ glibc-2.25/elf/rtld.c	2017-09-12 05:49:14.370807209 +0000
@@ -932,6 +932,18 @@
       main_map->l_name = (char *) "";
       *user_entry = main_map->l_entry;
 
+     /* GHUMVEE patch: ask the MVEE for the "virtualized" argv[0].  This can be
+     necessary if we run compile-time diversified variants that print out their
+     argv[0] value. */                                                                                                                                                                                             
+	 char virtualized_argv0[4096];
+	 if (syscall(MVEE_GET_VIRTUALIZED_ARGV0, _dl_argv[0], virtualized_argv0, 4096) == 0)
+	 {
+		 // should we do this?
+		 if (_dl_argv[0])
+			 free(_dl_argv[0]);
+		 _dl_argv[0] = strdup(virtualized_argv0);
+	 }
+
 #ifdef HAVE_AUX_VECTOR
       /* Adjust the on-stack auxiliary vector so that it looks like the
 	 binary was executed directly.  */
diff -N -r --unified glibc-2.25.orig/gmon/bb_init_func.c glibc-2.25/gmon/bb_init_func.c
--- glibc-2.25.orig/gmon/bb_init_func.c	2017-09-11 17:14:55.383218895 +0000
+++ glibc-2.25/gmon/bb_init_func.c	2017-09-11 22:18:57.825846466 +0000
@@ -26,6 +26,7 @@
 #include <sys/gmon.h>
 
 #include <stdlib.h>
+#include <atomic.h>
 
 void
 __bb_init_func (struct __bb *bb)
@@ -42,7 +43,7 @@
   bb->next = __bb_head;
   __bb_head = bb;
 
-  if (bb->next == 0 && p->state != GMON_PROF_ON)
+  if (bb->next == 0 && atomic_load(p->state) != GMON_PROF_ON)
     {
       /* we didn't register _mcleanup yet and pc profiling doesn't seem
 	 to be active, so let's register it now: */
diff -N -r --unified glibc-2.25.orig/gmon/gmon.c glibc-2.25/gmon/gmon.c
--- glibc-2.25.orig/gmon/gmon.c	2017-09-11 17:14:55.383218895 +0000
+++ glibc-2.25/gmon/gmon.c	2017-09-11 22:18:57.828846552 +0000
@@ -77,20 +77,20 @@
   struct gmonparam *p = &_gmonparam;
 
   /* Don't change the state if we ran into an error.  */
-  if (p->state == GMON_PROF_ERROR)
+  if (atomic_load(p->state) == GMON_PROF_ERROR)
     return;
 
   if (mode)
     {
       /* start */
       __profil((void *) p->kcount, p->kcountsize, p->lowpc, s_scale);
-      p->state = GMON_PROF_ON;
+      atomic_store(p->state, GMON_PROF_ON);
     }
   else
     {
       /* stop */
       __profil(NULL, 0, 0, 0);
-      p->state = GMON_PROF_OFF;
+      atomic_store(p->state, GMON_PROF_OFF);
     }
 }
 weak_alias (__moncontrol, moncontrol)
@@ -133,7 +133,7 @@
     {
       ERR("monstartup: out of memory\n");
       p->tos = NULL;
-      p->state = GMON_PROF_ERROR;
+      atomic_store(p->state, GMON_PROF_ERROR);
       return;
     }
   p->tos = (struct tostruct *)cp;
@@ -397,11 +397,11 @@
 void
 __write_profiling (void)
 {
-  int save = _gmonparam.state;
-  _gmonparam.state = GMON_PROF_OFF;
+  int save = atomic_load(_gmonparam.state);
+  atomic_store(_gmonparam.state, GMON_PROF_OFF);
   if (save == GMON_PROF_ON)
     write_gmon ();
-  _gmonparam.state = save;
+  atomic_store(_gmonparam.state, save);
 }
 #ifndef SHARED
 /* This symbol isn't used anywhere in the DSO and it is not exported.
@@ -418,7 +418,7 @@
 {
   __moncontrol (0);
 
-  if (_gmonparam.state != GMON_PROF_ERROR)
+  if (atomic_load(_gmonparam.state) != GMON_PROF_ERROR)
     write_gmon ();
 
   /* free the memory. */
diff -N -r --unified glibc-2.25.orig/gmon/mcount.c glibc-2.25/gmon/mcount.c
--- glibc-2.25.orig/gmon/mcount.c	2017-09-11 17:14:55.384218967 +0000
+++ glibc-2.25/gmon/mcount.c	2017-09-11 22:18:57.831846661 +0000
@@ -166,10 +166,10 @@
 
 	}
 done:
-	p->state = GMON_PROF_ON;
+	atomic_store(p->state, GMON_PROF_ON);
 	return;
 overflow:
-	p->state = GMON_PROF_ERROR;
+	atomic_store(p->state, GMON_PROF_ERROR);
 	return;
 }
 
diff -N -r --unified glibc-2.25.orig/include/atomic.h glibc-2.25/include/atomic.h
--- glibc-2.25.orig/include/atomic.h	2017-09-11 17:14:56.867325058 +0000
+++ glibc-2.25/include/atomic.h	2017-09-12 00:06:18.306197355 +0000
@@ -86,73 +86,73 @@
 
 /* Atomically store NEWVAL in *MEM if *MEM is equal to OLDVAL.
    Return the old *MEM value.  */
-#if !defined atomic_compare_and_exchange_val_acq \
-    && defined __arch_compare_and_exchange_val_32_acq
-# define atomic_compare_and_exchange_val_acq(mem, newval, oldval) \
-  __atomic_val_bysize (__arch_compare_and_exchange_val,acq,		      \
+#if !defined orig_atomic_compare_and_exchange_val_acq \
+    && defined orig___arch_compare_and_exchange_val_32_acq
+# define orig_atomic_compare_and_exchange_val_acq(mem, newval, oldval) \
+  __atomic_val_bysize (orig___arch_compare_and_exchange_val,acq,		      \
 		       mem, newval, oldval)
 #endif
 
 
-#ifndef catomic_compare_and_exchange_val_acq
-# ifdef __arch_c_compare_and_exchange_val_32_acq
-#  define catomic_compare_and_exchange_val_acq(mem, newval, oldval) \
-  __atomic_val_bysize (__arch_c_compare_and_exchange_val,acq,		      \
+#ifndef orig_catomic_compare_and_exchange_val_acq
+# ifdef orig___arch_c_compare_and_exchange_val_32_acq
+#  define orig_catomic_compare_and_exchange_val_acq(mem, newval, oldval) \
+  __atomic_val_bysize (orig___arch_c_compare_and_exchange_val,acq,		      \
 		       mem, newval, oldval)
 # else
-#  define catomic_compare_and_exchange_val_acq(mem, newval, oldval) \
-  atomic_compare_and_exchange_val_acq (mem, newval, oldval)
+#  define orig_catomic_compare_and_exchange_val_acq(mem, newval, oldval) \
+  orig_atomic_compare_and_exchange_val_acq (mem, newval, oldval)
 # endif
 #endif
 
 
-#ifndef catomic_compare_and_exchange_val_rel
-# ifndef atomic_compare_and_exchange_val_rel
-#  define catomic_compare_and_exchange_val_rel(mem, newval, oldval)	      \
-  catomic_compare_and_exchange_val_acq (mem, newval, oldval)
+#ifndef orig_catomic_compare_and_exchange_val_rel
+# ifndef orig_atomic_compare_and_exchange_val_rel
+#  define orig_catomic_compare_and_exchange_val_rel(mem, newval, oldval)	      \
+  orig_catomic_compare_and_exchange_val_acq (mem, newval, oldval)
 # else
-#  define catomic_compare_and_exchange_val_rel(mem, newval, oldval)	      \
-  atomic_compare_and_exchange_val_rel (mem, newval, oldval)
+#  define orig_catomic_compare_and_exchange_val_rel(mem, newval, oldval)	      \
+  orig_atomic_compare_and_exchange_val_rel (mem, newval, oldval)
 # endif
 #endif
 
 
-#ifndef atomic_compare_and_exchange_val_rel
-# define atomic_compare_and_exchange_val_rel(mem, newval, oldval)	      \
-  atomic_compare_and_exchange_val_acq (mem, newval, oldval)
+#ifndef orig_atomic_compare_and_exchange_val_rel
+# define orig_atomic_compare_and_exchange_val_rel(mem, newval, oldval)	      \
+  orig_atomic_compare_and_exchange_val_acq (mem, newval, oldval)
 #endif
 
 
 /* Atomically store NEWVAL in *MEM if *MEM is equal to OLDVAL.
    Return zero if *MEM was changed or non-zero if no exchange happened.  */
-#ifndef atomic_compare_and_exchange_bool_acq
-# ifdef __arch_compare_and_exchange_bool_32_acq
-#  define atomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
-  __atomic_bool_bysize (__arch_compare_and_exchange_bool,acq,		      \
+#ifndef orig_atomic_compare_and_exchange_bool_acq
+# ifdef orig___arch_compare_and_exchange_bool_32_acq
+#  define orig_atomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
+  __atomic_bool_bysize (orig___arch_compare_and_exchange_bool,acq,		      \
 		        mem, newval, oldval)
 # else
-#  define atomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
+#  define orig_atomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
   ({ /* Cannot use __oldval here, because macros later in this file might     \
 	call this macro with __oldval argument.	 */			      \
      __typeof (oldval) __atg3_old = (oldval);				      \
-     atomic_compare_and_exchange_val_acq (mem, newval, __atg3_old)	      \
+     orig_atomic_compare_and_exchange_val_acq (mem, newval, __atg3_old)	      \
        != __atg3_old;							      \
   })
 # endif
 #endif
 
 
-#ifndef catomic_compare_and_exchange_bool_acq
-# ifdef __arch_c_compare_and_exchange_bool_32_acq
-#  define catomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
-  __atomic_bool_bysize (__arch_c_compare_and_exchange_bool,acq,		      \
+#ifndef orig_catomic_compare_and_exchange_bool_acq
+# ifdef orig___arch_c_compare_and_exchange_bool_32_acq
+#  define orig_catomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
+  __atomic_bool_bysize (orig___arch_c_compare_and_exchange_bool,acq,		      \
 		        mem, newval, oldval)
 # else
-#  define catomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
+#  define orig_catomic_compare_and_exchange_bool_acq(mem, newval, oldval) \
   ({ /* Cannot use __oldval here, because macros later in this file might     \
 	call this macro with __oldval argument.	 */			      \
      __typeof (oldval) __atg4_old = (oldval);				      \
-     catomic_compare_and_exchange_val_acq (mem, newval, __atg4_old)	      \
+     orig_catomic_compare_and_exchange_val_acq (mem, newval, __atg4_old)	      \
        != __atg4_old;							      \
   })
 # endif
@@ -160,8 +160,8 @@
 
 
 /* Store NEWVALUE in *MEM and return the old value.  */
-#ifndef atomic_exchange_acq
-# define atomic_exchange_acq(mem, newvalue) \
+#ifndef orig_atomic_exchange_acq
+# define orig_atomic_exchange_acq(mem, newvalue) \
   ({ __typeof (*(mem)) __atg5_oldval;					      \
      __typeof (mem) __atg5_memp = (mem);				      \
      __typeof (*(mem)) __atg5_value = (newvalue);			      \
@@ -169,24 +169,24 @@
      do									      \
        __atg5_oldval = *__atg5_memp;					      \
      while (__builtin_expect						      \
-	    (atomic_compare_and_exchange_bool_acq (__atg5_memp, __atg5_value, \
+	    (orig_atomic_compare_and_exchange_bool_acq (__atg5_memp, __atg5_value, \
 						   __atg5_oldval), 0));	      \
 									      \
      __atg5_oldval; })
 #endif
 
-#ifndef atomic_exchange_rel
-# define atomic_exchange_rel(mem, newvalue) atomic_exchange_acq (mem, newvalue)
+#ifndef orig_atomic_exchange_rel
+# define orig_atomic_exchange_rel(mem, newvalue) orig_atomic_exchange_acq (mem, newvalue)
 #endif
 
 
 /* Add VALUE to *MEM and return the old value of *MEM.  */
-#ifndef atomic_exchange_and_add_acq
-# ifdef atomic_exchange_and_add
-#  define atomic_exchange_and_add_acq(mem, value) \
-  atomic_exchange_and_add (mem, value)
+#ifndef orig_atomic_exchange_and_add_acq
+# ifdef orig_atomic_exchange_and_add
+#  define orig_atomic_exchange_and_add_acq(mem, value) \
+  orig_atomic_exchange_and_add (mem, value)
 # else
-#  define atomic_exchange_and_add_acq(mem, value) \
+#  define orig_atomic_exchange_and_add_acq(mem, value) \
   ({ __typeof (*(mem)) __atg6_oldval;					      \
      __typeof (mem) __atg6_memp = (mem);				      \
      __typeof (*(mem)) __atg6_value = (value);				      \
@@ -194,7 +194,7 @@
      do									      \
        __atg6_oldval = *__atg6_memp;					      \
      while (__builtin_expect						      \
-	    (atomic_compare_and_exchange_bool_acq (__atg6_memp,		      \
+	    (orig_atomic_compare_and_exchange_bool_acq (__atg6_memp,		      \
 						   __atg6_oldval	      \
 						   + __atg6_value,	      \
 						   __atg6_oldval), 0));	      \
@@ -203,18 +203,18 @@
 # endif
 #endif
 
-#ifndef atomic_exchange_and_add_rel
-# define atomic_exchange_and_add_rel(mem, value) \
-  atomic_exchange_and_add_acq(mem, value)
+#ifndef orig_atomic_exchange_and_add_rel
+# define orig_atomic_exchange_and_add_rel(mem, value) \
+  orig_atomic_exchange_and_add_acq(mem, value)
 #endif
 
-#ifndef atomic_exchange_and_add
-# define atomic_exchange_and_add(mem, value) \
-  atomic_exchange_and_add_acq(mem, value)
+#ifndef orig_atomic_exchange_and_add
+# define orig_atomic_exchange_and_add(mem, value) \
+  orig_atomic_exchange_and_add_acq(mem, value)
 #endif
 
-#ifndef catomic_exchange_and_add
-# define catomic_exchange_and_add(mem, value) \
+#ifndef orig_catomic_exchange_and_add
+# define orig_catomic_exchange_and_add(mem, value) \
   ({ __typeof (*(mem)) __atg7_oldv;					      \
      __typeof (mem) __atg7_memp = (mem);				      \
      __typeof (*(mem)) __atg7_value = (value);				      \
@@ -222,7 +222,7 @@
      do									      \
        __atg7_oldv = *__atg7_memp;					      \
      while (__builtin_expect						      \
-	    (catomic_compare_and_exchange_bool_acq (__atg7_memp,	      \
+	    (orig_catomic_compare_and_exchange_bool_acq (__atg7_memp,	      \
 						    __atg7_oldv		      \
 						    + __atg7_value,	      \
 						    __atg7_oldv), 0));	      \
@@ -231,8 +231,8 @@
 #endif
 
 
-#ifndef atomic_max
-# define atomic_max(mem, value) \
+#ifndef orig_atomic_max
+# define orig_atomic_max(mem, value) \
   do {									      \
     __typeof (*(mem)) __atg8_oldval;					      \
     __typeof (mem) __atg8_memp = (mem);					      \
@@ -242,14 +242,14 @@
       if (__atg8_oldval >= __atg8_value)				      \
 	break;								      \
     } while (__builtin_expect						      \
-	     (atomic_compare_and_exchange_bool_acq (__atg8_memp, __atg8_value,\
+	     (orig_atomic_compare_and_exchange_bool_acq (__atg8_memp, __atg8_value,\
 						    __atg8_oldval), 0));      \
   } while (0)
 #endif
 
 
-#ifndef catomic_max
-# define catomic_max(mem, value) \
+#ifndef orig_catomic_max
+# define orig_catomic_max(mem, value) \
   do {									      \
     __typeof (*(mem)) __atg9_oldv;					      \
     __typeof (mem) __atg9_memp = (mem);					      \
@@ -259,15 +259,15 @@
       if (__atg9_oldv >= __atg9_value)					      \
 	break;								      \
     } while (__builtin_expect						      \
-	     (catomic_compare_and_exchange_bool_acq (__atg9_memp,	      \
+	     (orig_catomic_compare_and_exchange_bool_acq (__atg9_memp,	      \
 						     __atg9_value,	      \
 						     __atg9_oldv), 0));	      \
   } while (0)
 #endif
 
 
-#ifndef atomic_min
-# define atomic_min(mem, value) \
+#ifndef orig_atomic_min
+# define orig_atomic_min(mem, value) \
   do {									      \
     __typeof (*(mem)) __atg10_oldval;					      \
     __typeof (mem) __atg10_memp = (mem);				      \
@@ -277,81 +277,81 @@
       if (__atg10_oldval <= __atg10_value)				      \
 	break;								      \
     } while (__builtin_expect						      \
-	     (atomic_compare_and_exchange_bool_acq (__atg10_memp,	      \
+	     (orig_atomic_compare_and_exchange_bool_acq (__atg10_memp,	      \
 						    __atg10_value,	      \
 						    __atg10_oldval), 0));     \
   } while (0)
 #endif
 
 
-#ifndef atomic_add
-# define atomic_add(mem, value) (void) atomic_exchange_and_add ((mem), (value))
+#ifndef orig_atomic_add
+# define orig_atomic_add(mem, value) (void) orig_atomic_exchange_and_add ((mem), (value))
 #endif
 
 
-#ifndef catomic_add
-# define catomic_add(mem, value) \
-  (void) catomic_exchange_and_add ((mem), (value))
+#ifndef orig_catomic_add
+# define orig_catomic_add(mem, value) \
+  (void) orig_catomic_exchange_and_add ((mem), (value))
 #endif
 
 
-#ifndef atomic_increment
-# define atomic_increment(mem) atomic_add ((mem), 1)
+#ifndef orig_atomic_increment
+# define orig_atomic_increment(mem) orig_atomic_add ((mem), 1)
 #endif
 
 
-#ifndef catomic_increment
-# define catomic_increment(mem) catomic_add ((mem), 1)
+#ifndef orig_catomic_increment
+# define orig_catomic_increment(mem) orig_catomic_add ((mem), 1)
 #endif
 
 
-#ifndef atomic_increment_val
-# define atomic_increment_val(mem) (atomic_exchange_and_add ((mem), 1) + 1)
+#ifndef orig_atomic_increment_val
+# define orig_atomic_increment_val(mem) (orig_atomic_exchange_and_add ((mem), 1) + 1)
 #endif
 
 
-#ifndef catomic_increment_val
-# define catomic_increment_val(mem) (catomic_exchange_and_add ((mem), 1) + 1)
+#ifndef orig_catomic_increment_val
+# define orig_catomic_increment_val(mem) (orig_catomic_exchange_and_add ((mem), 1) + 1)
 #endif
 
 
 /* Add one to *MEM and return true iff it's now zero.  */
-#ifndef atomic_increment_and_test
-# define atomic_increment_and_test(mem) \
-  (atomic_exchange_and_add ((mem), 1) + 1 == 0)
+#ifndef orig_atomic_increment_and_test
+# define orig_atomic_increment_and_test(mem) \
+  (orig_atomic_exchange_and_add ((mem), 1) + 1 == 0)
 #endif
 
 
-#ifndef atomic_decrement
-# define atomic_decrement(mem) atomic_add ((mem), -1)
+#ifndef orig_atomic_decrement
+# define orig_atomic_decrement(mem) orig_atomic_add ((mem), -1)
 #endif
 
 
-#ifndef catomic_decrement
-# define catomic_decrement(mem) catomic_add ((mem), -1)
+#ifndef orig_catomic_decrement
+# define orig_catomic_decrement(mem) orig_catomic_add ((mem), -1)
 #endif
 
 
-#ifndef atomic_decrement_val
-# define atomic_decrement_val(mem) (atomic_exchange_and_add ((mem), -1) - 1)
+#ifndef orig_atomic_decrement_val
+# define orig_atomic_decrement_val(mem) (orig_atomic_exchange_and_add ((mem), -1) - 1)
 #endif
 
 
-#ifndef catomic_decrement_val
-# define catomic_decrement_val(mem) (catomic_exchange_and_add ((mem), -1) - 1)
+#ifndef orig_catomic_decrement_val
+# define orig_catomic_decrement_val(mem) (orig_catomic_exchange_and_add ((mem), -1) - 1)
 #endif
 
 
 /* Subtract 1 from *MEM and return true iff it's now zero.  */
-#ifndef atomic_decrement_and_test
-# define atomic_decrement_and_test(mem) \
-  (atomic_exchange_and_add ((mem), -1) == 1)
+#ifndef orig_atomic_decrement_and_test
+# define orig_atomic_decrement_and_test(mem) \
+  (orig_atomic_exchange_and_add ((mem), -1) == 1)
 #endif
 
 
 /* Decrement *MEM if it is > 0, and return the old value.  */
-#ifndef atomic_decrement_if_positive
-# define atomic_decrement_if_positive(mem) \
+#ifndef orig_atomic_decrement_if_positive
+# define orig_atomic_decrement_if_positive(mem) \
   ({ __typeof (*(mem)) __atg11_oldval;					      \
      __typeof (mem) __atg11_memp = (mem);				      \
 									      \
@@ -362,35 +362,35 @@
 	   break;							      \
        }								      \
      while (__builtin_expect						      \
-	    (atomic_compare_and_exchange_bool_acq (__atg11_memp,	      \
+	    (orig_atomic_compare_and_exchange_bool_acq (__atg11_memp,	      \
 						   __atg11_oldval - 1,	      \
 						   __atg11_oldval), 0));      \
      __atg11_oldval; })
 #endif
 
 
-#ifndef atomic_add_negative
-# define atomic_add_negative(mem, value)				      \
+#ifndef orig_atomic_add_negative
+# define orig_atomic_add_negative(mem, value)				      \
   ({ __typeof (value) __atg12_value = (value);				      \
-     atomic_exchange_and_add (mem, __atg12_value) < -__atg12_value; })
+     orig_atomic_exchange_and_add (mem, __atg12_value) < -__atg12_value; })
 #endif
 
 
-#ifndef atomic_add_zero
-# define atomic_add_zero(mem, value)					      \
+#ifndef orig_atomic_add_zero
+# define orig_atomic_add_zero(mem, value)					      \
   ({ __typeof (value) __atg13_value = (value);				      \
-     atomic_exchange_and_add (mem, __atg13_value) == -__atg13_value; })
+     orig_atomic_exchange_and_add (mem, __atg13_value) == -__atg13_value; })
 #endif
 
 
-#ifndef atomic_bit_set
-# define atomic_bit_set(mem, bit) \
-  (void) atomic_bit_test_set(mem, bit)
+#ifndef orig_atomic_bit_set
+# define orig_atomic_bit_set(mem, bit) \
+  (void) orig_atomic_bit_test_set(mem, bit)
 #endif
 
 
-#ifndef atomic_bit_test_set
-# define atomic_bit_test_set(mem, bit) \
+#ifndef orig_atomic_bit_test_set
+# define orig_atomic_bit_test_set(mem, bit) \
   ({ __typeof (*(mem)) __atg14_old;					      \
      __typeof (mem) __atg14_memp = (mem);				      \
      __typeof (*(mem)) __atg14_mask = ((__typeof (*(mem))) 1 << (bit));	      \
@@ -398,7 +398,7 @@
      do									      \
        __atg14_old = (*__atg14_memp);					      \
      while (__builtin_expect						      \
-	    (atomic_compare_and_exchange_bool_acq (__atg14_memp,	      \
+	    (orig_atomic_compare_and_exchange_bool_acq (__atg14_memp,	      \
 						   __atg14_old | __atg14_mask,\
 						   __atg14_old), 0));	      \
 									      \
@@ -406,8 +406,8 @@
 #endif
 
 /* Atomically *mem &= mask.  */
-#ifndef atomic_and
-# define atomic_and(mem, mask) \
+#ifndef orig_atomic_and
+# define orig_atomic_and(mem, mask) \
   do {									      \
     __typeof (*(mem)) __atg15_old;					      \
     __typeof (mem) __atg15_memp = (mem);				      \
@@ -416,14 +416,14 @@
     do									      \
       __atg15_old = (*__atg15_memp);					      \
     while (__builtin_expect						      \
-	   (atomic_compare_and_exchange_bool_acq (__atg15_memp,		      \
+	   (orig_atomic_compare_and_exchange_bool_acq (__atg15_memp,		      \
 						  __atg15_old & __atg15_mask, \
 						  __atg15_old), 0));	      \
   } while (0)
 #endif
 
-#ifndef catomic_and
-# define catomic_and(mem, mask) \
+#ifndef orig_catomic_and
+# define orig_catomic_and(mem, mask) \
   do {									      \
     __typeof (*(mem)) __atg20_old;					      \
     __typeof (mem) __atg20_memp = (mem);				      \
@@ -432,15 +432,15 @@
     do									      \
       __atg20_old = (*__atg20_memp);					      \
     while (__builtin_expect						      \
-	   (catomic_compare_and_exchange_bool_acq (__atg20_memp,	      \
+	   (orig_catomic_compare_and_exchange_bool_acq (__atg20_memp,	      \
 						   __atg20_old & __atg20_mask,\
 						   __atg20_old), 0));	      \
   } while (0)
 #endif
 
 /* Atomically *mem &= mask and return the old value of *mem.  */
-#ifndef atomic_and_val
-# define atomic_and_val(mem, mask) \
+#ifndef orig_atomic_and_val
+# define orig_atomic_and_val(mem, mask) \
   ({ __typeof (*(mem)) __atg16_old;					      \
      __typeof (mem) __atg16_memp = (mem);				      \
      __typeof (*(mem)) __atg16_mask = (mask);				      \
@@ -448,7 +448,7 @@
      do									      \
        __atg16_old = (*__atg16_memp);					      \
      while (__builtin_expect						      \
-	    (atomic_compare_and_exchange_bool_acq (__atg16_memp,	      \
+	    (orig_atomic_compare_and_exchange_bool_acq (__atg16_memp,	      \
 						   __atg16_old & __atg16_mask,\
 						   __atg16_old), 0));	      \
 									      \
@@ -456,8 +456,8 @@
 #endif
 
 /* Atomically *mem |= mask and return the old value of *mem.  */
-#ifndef atomic_or
-# define atomic_or(mem, mask) \
+#ifndef orig_atomic_or
+# define orig_atomic_or(mem, mask) \
   do {									      \
     __typeof (*(mem)) __atg17_old;					      \
     __typeof (mem) __atg17_memp = (mem);				      \
@@ -466,14 +466,14 @@
     do									      \
       __atg17_old = (*__atg17_memp);					      \
     while (__builtin_expect						      \
-	   (atomic_compare_and_exchange_bool_acq (__atg17_memp,		      \
+	   (orig_atomic_compare_and_exchange_bool_acq (__atg17_memp,		      \
 						  __atg17_old | __atg17_mask, \
 						  __atg17_old), 0));	      \
   } while (0)
 #endif
 
-#ifndef catomic_or
-# define catomic_or(mem, mask) \
+#ifndef orig_catomic_or
+# define orig_catomic_or(mem, mask) \
   do {									      \
     __typeof (*(mem)) __atg18_old;					      \
     __typeof (mem) __atg18_memp = (mem);				      \
@@ -482,15 +482,15 @@
     do									      \
       __atg18_old = (*__atg18_memp);					      \
     while (__builtin_expect						      \
-	   (catomic_compare_and_exchange_bool_acq (__atg18_memp,	      \
+	   (orig_catomic_compare_and_exchange_bool_acq (__atg18_memp,	      \
 						   __atg18_old | __atg18_mask,\
 						   __atg18_old), 0));	      \
   } while (0)
 #endif
 
 /* Atomically *mem |= mask and return the old value of *mem.  */
-#ifndef atomic_or_val
-# define atomic_or_val(mem, mask) \
+#ifndef orig_atomic_or_val
+# define orig_atomic_or_val(mem, mask) \
   ({ __typeof (*(mem)) __atg19_old;					      \
      __typeof (mem) __atg19_memp = (mem);				      \
      __typeof (*(mem)) __atg19_mask = (mask);				      \
@@ -498,7 +498,7 @@
      do									      \
        __atg19_old = (*__atg19_memp);					      \
      while (__builtin_expect						      \
-	    (atomic_compare_and_exchange_bool_acq (__atg19_memp,	      \
+	    (orig_atomic_compare_and_exchange_bool_acq (__atg19_memp,	      \
 						   __atg19_old | __atg19_mask,\
 						   __atg19_old), 0));	      \
 									      \
@@ -520,8 +520,8 @@
 #endif
 
 
-#ifndef atomic_forced_read
-# define atomic_forced_read(x) \
+#ifndef orig_atomic_forced_read
+# define orig_atomic_forced_read(x) \
   ({ __typeof (x) __x; __asm ("" : "=r" (__x) : "0" (x)); __x; })
 #endif
 
@@ -572,82 +572,82 @@
 # define atomic_thread_fence_seq_cst() \
   __atomic_thread_fence (__ATOMIC_SEQ_CST)
 
-# define atomic_load_relaxed(mem) \
+# define orig_atomic_load_relaxed(mem) \
   ({ __atomic_check_size_ls((mem));					      \
      __atomic_load_n ((mem), __ATOMIC_RELAXED); })
-# define atomic_load_acquire(mem) \
+# define orig_atomic_load_acquire(mem) \
   ({ __atomic_check_size_ls((mem));					      \
      __atomic_load_n ((mem), __ATOMIC_ACQUIRE); })
 
-# define atomic_store_relaxed(mem, val) \
+# define orig_atomic_store_relaxed(mem, val) \
   do {									      \
     __atomic_check_size_ls((mem));					      \
     __atomic_store_n ((mem), (val), __ATOMIC_RELAXED);			      \
   } while (0)
-# define atomic_store_release(mem, val) \
+# define orig_atomic_store_release(mem, val) \
   do {									      \
     __atomic_check_size_ls((mem));					      \
     __atomic_store_n ((mem), (val), __ATOMIC_RELEASE);			      \
   } while (0)
 
 /* On failure, this CAS has memory_order_relaxed semantics.  */
-# define atomic_compare_exchange_weak_relaxed(mem, expected, desired) \
+# define orig_atomic_compare_exchange_weak_relaxed(mem, expected, desired) \
   ({ __atomic_check_size((mem));					      \
   __atomic_compare_exchange_n ((mem), (expected), (desired), 1,		      \
     __ATOMIC_RELAXED, __ATOMIC_RELAXED); })
-# define atomic_compare_exchange_weak_acquire(mem, expected, desired) \
+# define orig_atomic_compare_exchange_weak_acquire(mem, expected, desired) \
   ({ __atomic_check_size((mem));					      \
   __atomic_compare_exchange_n ((mem), (expected), (desired), 1,		      \
     __ATOMIC_ACQUIRE, __ATOMIC_RELAXED); })
-# define atomic_compare_exchange_weak_release(mem, expected, desired) \
+# define orig_atomic_compare_exchange_weak_release(mem, expected, desired) \
   ({ __atomic_check_size((mem));					      \
   __atomic_compare_exchange_n ((mem), (expected), (desired), 1,		      \
     __ATOMIC_RELEASE, __ATOMIC_RELAXED); })
 
-# define atomic_exchange_relaxed(mem, desired) \
+# define orig_atomic_exchange_relaxed(mem, desired) \
   ({ __atomic_check_size((mem));					      \
   __atomic_exchange_n ((mem), (desired), __ATOMIC_RELAXED); })
-# define atomic_exchange_acquire(mem, desired) \
+# define orig_atomic_exchange_acquire(mem, desired) \
   ({ __atomic_check_size((mem));					      \
   __atomic_exchange_n ((mem), (desired), __ATOMIC_ACQUIRE); })
-# define atomic_exchange_release(mem, desired) \
+# define orig_atomic_exchange_release(mem, desired) \
   ({ __atomic_check_size((mem));					      \
   __atomic_exchange_n ((mem), (desired), __ATOMIC_RELEASE); })
 
-# define atomic_fetch_add_relaxed(mem, operand) \
+# define orig_atomic_fetch_add_relaxed(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_add ((mem), (operand), __ATOMIC_RELAXED); })
-# define atomic_fetch_add_acquire(mem, operand) \
+# define orig_atomic_fetch_add_acquire(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_add ((mem), (operand), __ATOMIC_ACQUIRE); })
-# define atomic_fetch_add_release(mem, operand) \
+# define orig_atomic_fetch_add_release(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_add ((mem), (operand), __ATOMIC_RELEASE); })
-# define atomic_fetch_add_acq_rel(mem, operand) \
+# define orig_atomic_fetch_add_acq_rel(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_add ((mem), (operand), __ATOMIC_ACQ_REL); })
 
-# define atomic_fetch_and_relaxed(mem, operand) \
+# define orig_atomic_fetch_and_relaxed(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_and ((mem), (operand), __ATOMIC_RELAXED); })
-# define atomic_fetch_and_acquire(mem, operand) \
+# define orig_atomic_fetch_and_acquire(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_and ((mem), (operand), __ATOMIC_ACQUIRE); })
-# define atomic_fetch_and_release(mem, operand) \
+# define orig_atomic_fetch_and_release(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_and ((mem), (operand), __ATOMIC_RELEASE); })
 
-# define atomic_fetch_or_relaxed(mem, operand) \
+# define orig_atomic_fetch_or_relaxed(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_or ((mem), (operand), __ATOMIC_RELAXED); })
-# define atomic_fetch_or_acquire(mem, operand) \
+# define orig_atomic_fetch_or_acquire(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_or ((mem), (operand), __ATOMIC_ACQUIRE); })
-# define atomic_fetch_or_release(mem, operand) \
+# define orig_atomic_fetch_or_release(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_or ((mem), (operand), __ATOMIC_RELEASE); })
 
-# define atomic_fetch_xor_release(mem, operand) \
+# define orig_atomic_fetch_xor_release(mem, operand) \
   ({ __atomic_check_size((mem));					      \
   __atomic_fetch_xor ((mem), (operand), __ATOMIC_RELEASE); })
 
@@ -666,25 +666,25 @@
 #  define atomic_thread_fence_seq_cst() atomic_full_barrier ()
 # endif
 
-# ifndef atomic_load_relaxed
-#  define atomic_load_relaxed(mem) \
+# ifndef orig_atomic_load_relaxed
+#  define orig_atomic_load_relaxed(mem) \
    ({ __typeof (*(mem)) __atg100_val;					      \
    __asm ("" : "=r" (__atg100_val) : "0" (*(mem)));			      \
    __atg100_val; })
 # endif
-# ifndef atomic_load_acquire
-#  define atomic_load_acquire(mem) \
+# ifndef orig_atomic_load_acquire
+#  define orig_atomic_load_acquire(mem) \
    ({ __typeof (*(mem)) __atg101_val = atomic_load_relaxed (mem);	      \
    atomic_thread_fence_acquire ();					      \
    __atg101_val; })
 # endif
 
-# ifndef atomic_store_relaxed
+# ifndef orig_atomic_store_relaxed
 /* XXX Use inline asm here?  */
-#  define atomic_store_relaxed(mem, val) do { *(mem) = (val); } while (0)
+#  define orig_atomic_store_relaxed(mem, val) do { *(mem) = (val); } while (0)
 # endif
-# ifndef atomic_store_release
-#  define atomic_store_release(mem, val) \
+# ifndef orig_atomic_store_release
+#  define orig_atomic_store_release(mem, val) \
    do {									      \
      atomic_thread_fence_release ();					      \
      atomic_store_relaxed ((mem), (val));				      \
@@ -695,107 +695,107 @@
 /* XXX This potentially has one branch more than necessary, but archs
    currently do not define a CAS that returns both the previous value and
    the success flag.  */
-# ifndef atomic_compare_exchange_weak_acquire
-#  define atomic_compare_exchange_weak_acquire(mem, expected, desired) \
+# ifndef orig_atomic_compare_exchange_weak_acquire
+#  define orig_atomic_compare_exchange_weak_acquire(mem, expected, desired) \
    ({ typeof (*(expected)) __atg102_expected = *(expected);		      \
    *(expected) =							      \
-     atomic_compare_and_exchange_val_acq ((mem), (desired), *(expected));     \
+     orig_atomic_compare_and_exchange_val_acq ((mem), (desired), *(expected));     \
    *(expected) == __atg102_expected; })
 # endif
-# ifndef atomic_compare_exchange_weak_relaxed
+# ifndef orig_atomic_compare_exchange_weak_relaxed
 /* XXX Fall back to CAS with acquire MO because archs do not define a weaker
    CAS.  */
-#  define atomic_compare_exchange_weak_relaxed(mem, expected, desired) \
-   atomic_compare_exchange_weak_acquire ((mem), (expected), (desired))
+#  define orig_atomic_compare_exchange_weak_relaxed(mem, expected, desired) \
+   orig_atomic_compare_exchange_weak_acquire ((mem), (expected), (desired))
 # endif
-# ifndef atomic_compare_exchange_weak_release
-#  define atomic_compare_exchange_weak_release(mem, expected, desired) \
+# ifndef orig_atomic_compare_exchange_weak_release
+#  define orig_atomic_compare_exchange_weak_release(mem, expected, desired) \
    ({ typeof (*(expected)) __atg103_expected = *(expected);		      \
    *(expected) =							      \
-     atomic_compare_and_exchange_val_rel ((mem), (desired), *(expected));     \
+     orig_atomic_compare_and_exchange_val_rel ((mem), (desired), *(expected));     \
    *(expected) == __atg103_expected; })
 # endif
 
 /* XXX Fall back to acquire MO because archs do not define a weaker
    atomic_exchange.  */
-# ifndef atomic_exchange_relaxed
-#  define atomic_exchange_relaxed(mem, val) \
-   atomic_exchange_acq ((mem), (val))
+# ifndef orig_atomic_exchange_relaxed
+#  define orig_atomic_exchange_relaxed(mem, val) \
+   orig_atomic_exchange_acq ((mem), (val))
 # endif
-# ifndef atomic_exchange_acquire
-#  define atomic_exchange_acquire(mem, val) \
-   atomic_exchange_acq ((mem), (val))
+# ifndef orig_atomic_exchange_acquire
+#  define orig_atomic_exchange_acquire(mem, val) \
+   orig_atomic_exchange_acq ((mem), (val))
 # endif
-# ifndef atomic_exchange_release
-#  define atomic_exchange_release(mem, val) \
-   atomic_exchange_rel ((mem), (val))
+# ifndef orig_atomic_exchange_release
+#  define orig_atomic_exchange_release(mem, val) \
+   orig_atomic_exchange_rel ((mem), (val))
 # endif
 
-# ifndef atomic_fetch_add_acquire
-#  define atomic_fetch_add_acquire(mem, operand) \
-   atomic_exchange_and_add_acq ((mem), (operand))
+# ifndef orig_atomic_fetch_add_acquire
+#  define orig_atomic_fetch_add_acquire(mem, operand) \
+   orig_atomic_exchange_and_add_acq ((mem), (operand))
 # endif
-# ifndef atomic_fetch_add_relaxed
+# ifndef orig_atomic_fetch_add_relaxed
 /* XXX Fall back to acquire MO because the MO semantics of
    atomic_exchange_and_add are not documented; the generic version falls back
    to atomic_exchange_and_add_acq if atomic_exchange_and_add is not defined,
    and vice versa.  */
-#  define atomic_fetch_add_relaxed(mem, operand) \
-   atomic_fetch_add_acquire ((mem), (operand))
+#  define orig_atomic_fetch_add_relaxed(mem, operand) \
+   orig_atomic_fetch_add_acquire ((mem), (operand))
 # endif
-# ifndef atomic_fetch_add_release
-#  define atomic_fetch_add_release(mem, operand) \
-   atomic_exchange_and_add_rel ((mem), (operand))
+# ifndef orig_atomic_fetch_add_release
+#  define orig_atomic_fetch_add_release(mem, operand) \
+   orig_atomic_exchange_and_add_rel ((mem), (operand))
 # endif
-# ifndef atomic_fetch_add_acq_rel
-#  define atomic_fetch_add_acq_rel(mem, operand) \
+# ifndef orig_atomic_fetch_add_acq_rel
+#  define orig_atomic_fetch_add_acq_rel(mem, operand) \
    ({ atomic_thread_fence_release ();					      \
-   atomic_exchange_and_add_acq ((mem), (operand)); })
+   orig_atomic_exchange_and_add_acq ((mem), (operand)); })
 # endif
 
 /* XXX Fall back to acquire MO because archs do not define a weaker
    atomic_and_val.  */
-# ifndef atomic_fetch_and_relaxed
-#  define atomic_fetch_and_relaxed(mem, operand) \
-   atomic_fetch_and_acquire ((mem), (operand))
+# ifndef orig_atomic_fetch_and_relaxed
+#  define orig_atomic_fetch_and_relaxed(mem, operand) \
+   orig_atomic_fetch_and_acquire ((mem), (operand))
 # endif
 /* XXX The default for atomic_and_val has acquire semantics, but this is not
    documented.  */
-# ifndef atomic_fetch_and_acquire
-#  define atomic_fetch_and_acquire(mem, operand) \
-   atomic_and_val ((mem), (operand))
+# ifndef orig_atomic_fetch_and_acquire
+#  define orig_atomic_fetch_and_acquire(mem, operand) \
+   orig_atomic_and_val ((mem), (operand))
 # endif
-# ifndef atomic_fetch_and_release
+# ifndef orig_atomic_fetch_and_release
 /* XXX This unnecessarily has acquire MO.  */
-#  define atomic_fetch_and_release(mem, operand) \
+#  define orig_atomic_fetch_and_release(mem, operand) \
    ({ atomic_thread_fence_release ();					      \
-   atomic_and_val ((mem), (operand)); })
+   orig_atomic_and_val ((mem), (operand)); })
 # endif
 
 /* XXX The default for atomic_or_val has acquire semantics, but this is not
    documented.  */
-# ifndef atomic_fetch_or_acquire
-#  define atomic_fetch_or_acquire(mem, operand) \
-   atomic_or_val ((mem), (operand))
+# ifndef orig_atomic_fetch_or_acquire
+#  define orig_atomic_fetch_or_acquire(mem, operand) \
+   orig_atomic_or_val ((mem), (operand))
 # endif
 /* XXX Fall back to acquire MO because archs do not define a weaker
    atomic_or_val.  */
-# ifndef atomic_fetch_or_relaxed
-#  define atomic_fetch_or_relaxed(mem, operand) \
-   atomic_fetch_or_acquire ((mem), (operand))
+# ifndef orig_atomic_fetch_or_relaxed
+#  define orig_atomic_fetch_or_relaxed(mem, operand) \
+   orig_atomic_fetch_or_acquire ((mem), (operand))
 # endif
 /* XXX Contains an unnecessary acquire MO because archs do not define a weaker
    atomic_or_val.  */
-# ifndef atomic_fetch_or_release
-#  define atomic_fetch_or_release(mem, operand) \
+# ifndef orig_atomic_fetch_or_release
+#  define orig_atomic_fetch_or_release(mem, operand) \
    ({ atomic_thread_fence_release ();					      \
-   atomic_fetch_or_acquire ((mem), (operand)); })
+   orig_atomic_fetch_or_acquire ((mem), (operand)); })
 # endif
 
-# ifndef atomic_fetch_xor_release
+# ifndef orig_atomic_fetch_xor_release
 /* Failing the atomic_compare_exchange_weak_release reloads the value in
    __atg104_expected, so we need only do the XOR again and retry.  */
-# define atomic_fetch_xor_release(mem, operand) \
+# define orig_atomic_fetch_xor_release(mem, operand) \
   ({ __typeof (mem) __atg104_memp = (mem);				      \
      __typeof (*(mem)) __atg104_expected = (*__atg104_memp);		      \
      __typeof (*(mem)) __atg104_desired;				      \
@@ -804,7 +804,7 @@
      do									      \
        __atg104_desired = __atg104_expected ^ __atg104_op;		      \
      while (__glibc_unlikely						      \
-	    (atomic_compare_exchange_weak_release (			      \
+	    (orig_atomic_compare_exchange_weak_release (			      \
 	       __atg104_memp, &__atg104_expected, __atg104_desired)	      \
 	     == 0));							      \
      __atg104_expected; })
diff -N -r --unified glibc-2.25.orig/inet/getnameinfo.c glibc-2.25/inet/getnameinfo.c
--- glibc-2.25.orig/inet/getnameinfo.c	2017-09-11 17:14:56.983333356 +0000
+++ glibc-2.25/inet/getnameinfo.c	2017-09-11 22:18:57.849847554 +0000
@@ -91,8 +91,6 @@
 {
   static int not_first;
 
-  if (! not_first)
-    {
       __libc_lock_define_initialized (static, lock);
       __libc_lock_lock (lock);
 
@@ -182,7 +180,6 @@
 	}
 
       __libc_lock_unlock (lock);
-    }
 
   return domain;
 };
diff -N -r --unified glibc-2.25.orig/malloc/arena.c glibc-2.25/malloc/arena.c
--- glibc-2.25.orig/malloc/arena.c	2017-09-11 17:15:01.342645189 +0000
+++ glibc-2.25/malloc/arena.c	2017-09-11 22:18:57.855847910 +0000
@@ -456,7 +456,7 @@
 new_heap (size_t size, size_t top_pad)
 {
   size_t pagesize = GLRO (dl_pagesize);
-  char *p1, *p2;
+  char *p1, *p2, *prev_heap_area;
   unsigned long ul;
   heap_info *h;
 
@@ -475,12 +475,13 @@
      mapping (on Linux, this is the case for all non-writable mappings
      anyway). */
   p2 = MAP_FAILED;
-  if (aligned_heap_area)
+  prev_heap_area = atomic_load(aligned_heap_area);
+  if (prev_heap_area)
     {
-      p2 = (char *) MMAP (aligned_heap_area, HEAP_MAX_SIZE, PROT_NONE,
+      p2 = (char *) MMAP (prev_heap_area, HEAP_MAX_SIZE, PROT_NONE,
                           MAP_NORESERVE);
-      aligned_heap_area = NULL;
-      if (p2 != MAP_FAILED && ((unsigned long) p2 & (HEAP_MAX_SIZE - 1)))
+      atomic_store(aligned_heap_area, NULL);
+      if (p2 != MAP_FAILED && !mvee_all_heaps_aligned(p2, HEAP_MAX_SIZE))
         {
           __munmap (p2, HEAP_MAX_SIZE);
           p2 = MAP_FAILED;
@@ -488,6 +489,7 @@
     }
   if (p2 == MAP_FAILED)
     {
+	  (void) mvee_all_heaps_aligned(0, HEAP_MAX_SIZE);
       p1 = (char *) MMAP (0, HEAP_MAX_SIZE << 1, PROT_NONE, MAP_NORESERVE);
       if (p1 != MAP_FAILED)
         {
@@ -496,9 +498,8 @@
           ul = p2 - p1;
           if (ul)
             __munmap (p1, ul);
-          else
-            aligned_heap_area = p2 + HEAP_MAX_SIZE;
           __munmap (p2 + HEAP_MAX_SIZE, HEAP_MAX_SIZE - ul);
+		  atomic_store(aligned_heap_area, p2 + HEAP_MAX_SIZE);
         }
       else
         {
@@ -508,7 +509,7 @@
           if (p2 == MAP_FAILED)
             return 0;
 
-          if ((unsigned long) p2 & (HEAP_MAX_SIZE - 1))
+          if (!mvee_all_heaps_aligned(p2, HEAP_MAX_SIZE))
             {
               __munmap (p2, HEAP_MAX_SIZE);
               return 0;
@@ -590,8 +591,8 @@
 
 #define delete_heap(heap) \
   do {									      \
-      if ((char *) (heap) + HEAP_MAX_SIZE == aligned_heap_area)		      \
-        aligned_heap_area = NULL;					      \
+	if ((char *) (heap) + HEAP_MAX_SIZE == atomic_load(aligned_heap_area)) \
+	  atomic_store(aligned_heap_area, NULL);							\
       __munmap ((char *) (heap), HEAP_MAX_SIZE);			      \
     } while (0)
 
@@ -901,7 +902,7 @@
         {
           if (mp_.arena_max != 0)
             narenas_limit = mp_.arena_max;
-          else if (narenas > mp_.arena_test)
+          else if (atomic_load(narenas) > mp_.arena_test)
             {
               int n = __get_nprocs ();
 
@@ -914,7 +915,7 @@
             }
         }
     repeat:;
-      size_t n = narenas;
+      size_t n = atomic_load(narenas);
       /* NB: the following depends on the fact that (size_t)0 - 1 is a
          very large number and that the underflow is OK.  If arena_max
          is set the value of arena_test is irrelevant.  If arena_test
diff -N -r --unified glibc-2.25.orig/malloc/hooks.c glibc-2.25/malloc/hooks.c
--- glibc-2.25.orig/malloc/hooks.c	2017-09-11 17:15:01.361646548 +0000
+++ glibc-2.25/malloc/hooks.c	2017-09-11 22:18:57.858848122 +0000
@@ -77,10 +77,10 @@
       return;
     }
   using_malloc_checking = 1;
-  __malloc_hook = malloc_check;
-  __free_hook = free_check;
-  __realloc_hook = realloc_check;
-  __memalign_hook = memalign_check;
+  atomic_store(__malloc_hook, malloc_check);
+  atomic_store(__free_hook, free_check);
+  atomic_store(__realloc_hook, realloc_check);
+  atomic_store(__memalign_hook, memalign_check);
 }
 
 /* A simple, standard set of debugging hooks.  Overhead is `only' one
@@ -525,10 +525,10 @@
      cannot be more than one thread when we reach this point.  */
 
   /* Disable the malloc hooks (and malloc checking).  */
-  __malloc_hook = NULL;
-  __realloc_hook = NULL;
-  __free_hook = NULL;
-  __memalign_hook = NULL;
+  atomic_store(__malloc_hook, NULL);
+  atomic_store(__realloc_hook, NULL);
+  atomic_store(__free_hook, NULL);
+  atomic_store(__memalign_hook, NULL);
   using_malloc_checking = 0;
 
   /* Patch the dumped heap.  We no longer try to integrate into the
diff -N -r --unified glibc-2.25.orig/malloc/malloc.c glibc-2.25/malloc/malloc.c
--- glibc-2.25.orig/malloc/malloc.c	2017-09-11 17:15:01.384648194 +0000
+++ glibc-2.25/malloc/malloc.c	2017-09-11 22:18:57.874849278 +0000
@@ -1601,7 +1601,7 @@
 
 #define FASTCHUNKS_BIT        (1U)
 
-#define have_fastchunks(M)     (((M)->flags & FASTCHUNKS_BIT) == 0)
+#define have_fastchunks(M)     ((atomic_load((M)->flags) & FASTCHUNKS_BIT) == 0)
 #define clear_fastchunks(M)    catomic_or (&(M)->flags, FASTCHUNKS_BIT)
 #define set_fastchunks(M)      catomic_and (&(M)->flags, ~FASTCHUNKS_BIT)
 
@@ -2120,7 +2120,7 @@
 
   for (i = 0; i < NFASTBINS; ++i)
     {
-      p = fastbin (av, i);
+      p = atomic_load(fastbin (av, i));
 
       /* The following test can only be performed for the main arena.
          While mallopt calls malloc_consolidate to get rid of all fast
@@ -2272,7 +2272,7 @@
 
   if (av == NULL
       || ((unsigned long) (nb) >= (unsigned long) (mp_.mmap_threshold)
-	  && (mp_.n_mmaps < mp_.n_mmaps_max)))
+		  && (atomic_load(mp_.n_mmaps) < mp_.n_mmaps_max)))
     {
       char *mm;           /* return value from mmap call*/
 
@@ -3368,7 +3368,7 @@
     {
       idx = fastbin_index (nb);
       mfastbinptr *fb = &fastbin (av, idx);
-      mchunkptr pp = *fb;
+      mchunkptr pp = atomic_load(*fb);
       do
         {
           victim = pp;
@@ -3932,7 +3932,7 @@
     fb = &fastbin (av, idx);
 
     /* Atomically link P to its fastbin: P->FD = *FB; *FB = P;  */
-    mchunkptr old = *fb, old2;
+    mchunkptr old = atomic_load(*fb), old2;
     unsigned int old_idx = ~0u;
     do
       {
@@ -4648,7 +4648,7 @@
 
   for (i = 0; i < NFASTBINS; ++i)
     {
-      for (p = fastbin (av, i); p != 0; p = p->fd)
+      for (p = atomic_load(fastbin (av, i)); p != 0; p = p->fd)
         {
           ++nfastblocks;
           fastavail += chunksize (p);
@@ -4676,8 +4676,8 @@
   m->fsmblks += fastavail;
   if (av == &main_arena)
     {
-      m->hblks = mp_.n_mmaps;
-      m->hblkhd = mp_.mmapped_mem;
+      m->hblks = atomic_load(mp_.n_mmaps);
+      m->hblkhd = atomic_load(mp_.mmapped_mem);
       m->usmblks = 0;
       m->keepcost = chunksize (av->top);
     }
@@ -4717,7 +4717,7 @@
 {
   int i;
   mstate ar_ptr;
-  unsigned int in_use_b = mp_.mmapped_mem, system_b = in_use_b;
+  unsigned int in_use_b = atomic_load(mp_.mmapped_mem), system_b = in_use_b;
 
   if (__malloc_initialized < 0)
     ptmalloc_init ();
@@ -4748,9 +4748,9 @@
   fprintf (stderr, "Total (incl. mmap):\n");
   fprintf (stderr, "system bytes     = %10u\n", system_b);
   fprintf (stderr, "in use bytes     = %10u\n", in_use_b);
-  fprintf (stderr, "max mmap regions = %10u\n", (unsigned int) mp_.max_n_mmaps);
+  fprintf (stderr, "max mmap regions = %10u\n", (unsigned int) atomic_load(mp_.max_n_mmaps));
   fprintf (stderr, "max mmap bytes   = %10lu\n",
-           (unsigned long) mp_.max_mmapped_mem);
+           (unsigned long) atomic_load(mp_.max_mmapped_mem));
   ((_IO_FILE *) stderr)->_flags2 |= old_flags2;
   _IO_funlockfile (stderr);
 }
@@ -5156,7 +5156,7 @@
 
       for (size_t i = 0; i < NFASTBINS; ++i)
 	{
-	  mchunkptr p = fastbin (ar_ptr, i);
+	  mchunkptr p = atomic_load(fastbin (ar_ptr, i));
 	  if (p != NULL)
 	    {
 	      size_t nthissize = 0;
@@ -5278,7 +5278,7 @@
 	   "<aspace type=\"mprotect\" size=\"%zu\"/>\n"
 	   "</malloc>\n",
 	   total_nfastblocks, total_fastavail, total_nblocks, total_avail,
-	   mp_.n_mmaps, mp_.mmapped_mem,
+		   atomic_load(mp_.n_mmaps), atomic_load(mp_.mmapped_mem),
 	   total_system, total_max_system,
 	   total_aspace, total_aspace_mprotect);
 
diff -N -r --unified glibc-2.25.orig/malloc/mcheck.c glibc-2.25/malloc/mcheck.c
--- glibc-2.25.orig/malloc/mcheck.c	2017-09-11 17:15:01.386648337 +0000
+++ glibc-2.25/malloc/mcheck.c	2017-09-11 22:18:57.879849638 +0000
@@ -189,12 +189,12 @@
       flood (ptr, FREEFLOOD, hdr->size);
       ptr = hdr->block;
     }
-  __free_hook = old_free_hook;
+  atomic_store(__free_hook, old_free_hook);
   if (old_free_hook != NULL)
     (*old_free_hook)(ptr, caller);
   else
     free (ptr);
-  __free_hook = freehook;
+  atomic_store(__free_hook, freehook);
 }
 
 static __ptr_t
@@ -211,13 +211,13 @@
       return NULL;
     }
 
-  __malloc_hook = old_malloc_hook;
+  atomic_store(__malloc_hook, old_malloc_hook);
   if (old_malloc_hook != NULL)
     hdr = (struct hdr *) (*old_malloc_hook)(sizeof (struct hdr) + size + 1,
                                             caller);
   else
     hdr = (struct hdr *) malloc (sizeof (struct hdr) + size + 1);
-  __malloc_hook = mallochook;
+  atomic_store(__malloc_hook, mallochook);
   if (hdr == NULL)
     return NULL;
 
@@ -249,12 +249,12 @@
       return NULL;
     }
 
-  __memalign_hook = old_memalign_hook;
+  atomic_store(__memalign_hook, old_memalign_hook);
   if (old_memalign_hook != NULL)
     block = (*old_memalign_hook)(alignment, slop + size + 1, caller);
   else
     block = memalign (alignment, slop + size + 1);
-  __memalign_hook = memalignhook;
+  atomic_store(__memalign_hook, memalignhook);
   if (block == NULL)
     return NULL;
 
@@ -305,10 +305,10 @@
       osize = 0;
       hdr = NULL;
     }
-  __free_hook = old_free_hook;
-  __malloc_hook = old_malloc_hook;
-  __memalign_hook = old_memalign_hook;
-  __realloc_hook = old_realloc_hook;
+  atomic_store(__free_hook, old_free_hook);
+  atomic_store(__malloc_hook, old_malloc_hook);
+  atomic_store(__memalign_hook, old_memalign_hook);
+  atomic_store(__realloc_hook, old_realloc_hook);
   if (old_realloc_hook != NULL)
     hdr = (struct hdr *) (*old_realloc_hook)((__ptr_t) hdr,
                                              sizeof (struct hdr) + size + 1,
@@ -316,10 +316,10 @@
   else
     hdr = (struct hdr *) realloc ((__ptr_t) hdr,
                                   sizeof (struct hdr) + size + 1);
-  __free_hook = freehook;
-  __malloc_hook = mallochook;
-  __memalign_hook = memalignhook;
-  __realloc_hook = reallochook;
+  atomic_store(__free_hook, freehook);
+  atomic_store(__malloc_hook, mallochook);
+  atomic_store(__memalign_hook, memalignhook);
+  atomic_store(__realloc_hook, reallochook);
   if (hdr == NULL)
     return NULL;
 
@@ -383,15 +383,15 @@
       p = malloc_opt_barrier (p);
       free (p);
 
-      old_free_hook = __free_hook;
-      __free_hook = freehook;
-      old_malloc_hook = __malloc_hook;
-      __malloc_hook = mallochook;
-      old_memalign_hook = __memalign_hook;
-      __memalign_hook = memalignhook;
-      old_realloc_hook = __realloc_hook;
-      __realloc_hook = reallochook;
-      mcheck_used = 1;
+      old_free_hook = atomic_load(__free_hook);
+      atomic_store(__free_hook, freehook);
+      old_malloc_hook = atomic_load(__malloc_hook);
+      atomic_store(__malloc_hook, mallochook);
+      old_memalign_hook = atomic_load(__memalign_hook);
+      atomic_store(__memalign_hook, memalignhook);
+      old_realloc_hook = atomic_load(__realloc_hook);
+	  atomic_store(__realloc_hook, reallochook);      
+	  mcheck_used = 1;
     }
 
   return mcheck_used ? 0 : -1;
diff -N -r --unified glibc-2.25.orig/malloc/mtrace.c glibc-2.25/malloc/mtrace.c
--- glibc-2.25.orig/malloc/mtrace.c	2017-09-11 17:15:01.407649839 +0000
+++ glibc-2.25/malloc/mtrace.c	2017-09-11 22:18:57.884849997 +0000
@@ -137,12 +137,12 @@
       tr_break ();
       __libc_lock_lock (lock);
     }
-  __free_hook = tr_old_free_hook;
+  atomic_store(__free_hook, tr_old_free_hook);
   if (tr_old_free_hook != NULL)
     (*tr_old_free_hook)(ptr, caller);
   else
     free (ptr);
-  __free_hook = tr_freehook;
+  atomic_store(__free_hook, tr_freehook);
   __libc_lock_unlock (lock);
 }
 
@@ -154,12 +154,12 @@
   Dl_info mem;
   Dl_info *info = lock_and_info (caller, &mem);
 
-  __malloc_hook = tr_old_malloc_hook;
+  atomic_store(__malloc_hook, tr_old_malloc_hook);
   if (tr_old_malloc_hook != NULL)
     hdr = (__ptr_t) (*tr_old_malloc_hook)(size, caller);
   else
     hdr = (__ptr_t) malloc (size);
-  __malloc_hook = tr_mallochook;
+  atomic_store(__malloc_hook, tr_mallochook);
 
   tr_where (caller, info);
   /* We could be printing a NULL here; that's OK.  */
@@ -184,16 +184,16 @@
   Dl_info mem;
   Dl_info *info = lock_and_info (caller, &mem);
 
-  __free_hook = tr_old_free_hook;
-  __malloc_hook = tr_old_malloc_hook;
-  __realloc_hook = tr_old_realloc_hook;
+  atomic_store(__free_hook, tr_old_free_hook);
+  atomic_store(__malloc_hook, tr_old_malloc_hook);
+  atomic_store(__realloc_hook, tr_old_realloc_hook);
   if (tr_old_realloc_hook != NULL)
     hdr = (__ptr_t) (*tr_old_realloc_hook)(ptr, size, caller);
   else
     hdr = (__ptr_t) realloc (ptr, size);
-  __free_hook = tr_freehook;
-  __malloc_hook = tr_mallochook;
-  __realloc_hook = tr_reallochook;
+  atomic_store(__free_hook, tr_freehook);
+  atomic_store(__malloc_hook, tr_mallochook);
+  atomic_store(__realloc_hook, tr_reallochook);
 
   tr_where (caller, info);
   if (hdr == NULL)
@@ -229,14 +229,14 @@
   Dl_info mem;
   Dl_info *info = lock_and_info (caller, &mem);
 
-  __memalign_hook = tr_old_memalign_hook;
-  __malloc_hook = tr_old_malloc_hook;
+  atomic_store(__memalign_hook, tr_old_memalign_hook);
+  atomic_store(__malloc_hook, tr_old_malloc_hook);
   if (tr_old_memalign_hook != NULL)
     hdr = (__ptr_t) (*tr_old_memalign_hook)(alignment, size, caller);
   else
     hdr = (__ptr_t) memalign (alignment, size);
-  __memalign_hook = tr_memalignhook;
-  __malloc_hook = tr_mallochook;
+  atomic_store(__memalign_hook, tr_memalignhook);
+  atomic_store(__malloc_hook, tr_mallochook);
 
   tr_where (caller, info);
   /* We could be printing a NULL here; that's OK.  */
@@ -313,14 +313,14 @@
           malloc_trace_buffer = mtb;
           setvbuf (mallstream, malloc_trace_buffer, _IOFBF, TRACE_BUFFER_SIZE);
           fprintf (mallstream, "= Start\n");
-          tr_old_free_hook = __free_hook;
-          __free_hook = tr_freehook;
-          tr_old_malloc_hook = __malloc_hook;
-          __malloc_hook = tr_mallochook;
-          tr_old_realloc_hook = __realloc_hook;
-          __realloc_hook = tr_reallochook;
-          tr_old_memalign_hook = __memalign_hook;
-          __memalign_hook = tr_memalignhook;
+          tr_old_free_hook = atomic_load(__free_hook);
+          atomic_store(__free_hook, tr_freehook);
+          tr_old_malloc_hook = atomic_load(__malloc_hook);
+          atomic_store(__malloc_hook, tr_mallochook);
+          tr_old_realloc_hook = atomic_load(__realloc_hook);
+          atomic_store(__realloc_hook, tr_reallochook);
+          tr_old_memalign_hook = atomic_load(__memalign_hook);
+          atomic_store(__memalign_hook, tr_memalignhook);
 #ifdef _LIBC
           if (!added_atexit_handler)
             {
@@ -347,10 +347,10 @@
      file.  */
   FILE *f = mallstream;
   mallstream = NULL;
-  __free_hook = tr_old_free_hook;
-  __malloc_hook = tr_old_malloc_hook;
-  __realloc_hook = tr_old_realloc_hook;
-  __memalign_hook = tr_old_memalign_hook;
+  atomic_store(__free_hook, tr_old_free_hook);
+  atomic_store(__malloc_hook, tr_old_malloc_hook);
+  atomic_store(__realloc_hook, tr_old_realloc_hook);
+  atomic_store(__memalign_hook, tr_old_memalign_hook);
 
   fprintf (f, "= End\n");
   fclose (f);
diff -N -r --unified glibc-2.25.orig/nptl/allocatestack.c glibc-2.25/nptl/allocatestack.c
--- glibc-2.25.orig/nptl/allocatestack.c	2017-09-11 17:15:06.345003021 +0000
+++ glibc-2.25/nptl/allocatestack.c	2017-09-12 07:10:27.055284719 +0000
@@ -865,7 +865,7 @@
       if (curp != self)
 	{
 	  /* This marks the stack as free.  */
-	  curp->tid = 0;
+		atomic_store(curp->tid, 0);
 
 	  /* Account for the size of the stack.  */
 	  stack_cache_actsize += curp->stackblock_size;
diff -N -r --unified glibc-2.25.orig/nptl/cancellation.c glibc-2.25/nptl/cancellation.c
--- glibc-2.25.orig/nptl/cancellation.c	2017-09-11 17:15:06.346003093 +0000
+++ glibc-2.25/nptl/cancellation.c	2017-09-11 22:18:57.886850141 +0000
@@ -30,7 +30,7 @@
 __pthread_enable_asynccancel (void)
 {
   struct pthread *self = THREAD_SELF;
-  int oldval = THREAD_GETMEM (self, cancelhandling);
+  int oldval = THREAD_ATOMIC_GETMEM (self, cancelhandling);
 
   while (1)
     {
@@ -45,7 +45,7 @@
 	{
 	  if (CANCEL_ENABLED_AND_CANCELED_AND_ASYNCHRONOUS (newval))
 	    {
-	      THREAD_SETMEM (self, result, PTHREAD_CANCELED);
+	      THREAD_ATOMIC_SETMEM (self, result, PTHREAD_CANCELED);
 	      __do_cancel ();
 	    }
 
@@ -72,7 +72,7 @@
   struct pthread *self = THREAD_SELF;
   int newval;
 
-  int oldval = THREAD_GETMEM (self, cancelhandling);
+  int oldval = THREAD_ATOMIC_GETMEM (self, cancelhandling);
 
   while (1)
     {
@@ -96,6 +96,6 @@
     {
       futex_wait_simple ((unsigned int *) &self->cancelhandling, newval,
 			 FUTEX_PRIVATE);
-      newval = THREAD_GETMEM (self, cancelhandling);
+      newval = THREAD_ATOMIC_GETMEM (self, cancelhandling);
     }
 }
diff -N -r --unified glibc-2.25.orig/nptl/lll_timedwait_tid.c glibc-2.25/nptl/lll_timedwait_tid.c
--- glibc-2.25.orig/nptl/lll_timedwait_tid.c	2017-09-11 17:15:06.360004095 +0000
+++ glibc-2.25/nptl/lll_timedwait_tid.c	2017-09-12 07:32:50.371095975 +0000
@@ -37,7 +37,8 @@
     return EINVAL;
 
   /* Repeat until thread terminated.  */
-  while ((tid = *tidp) != 0)
+  while ((tid = atomic_load(*tidp)) != 0 ||
+		 mvee_should_sync_tid())
     {
       struct timeval tv;
       struct timespec rt;
@@ -62,7 +63,8 @@
          The kernel up to version 3.16.3 does not use the private futex
          operations for futex wake-up when the clone terminates.
       */
-      if (lll_futex_timed_wait (tidp, tid, &rt, LLL_SHARED) == -ETIMEDOUT)
+	  if (lll_futex_syscall (4, tidp, __lll_private_flag(mvee_should_sync_tid() ? MVEE_FUTEX_WAIT_TID : FUTEX_WAIT, LLL_SHARED),
+							 tid, &rt) == -ETIMEDOUT)
         return ETIMEDOUT;
     }
 
diff -N -r --unified glibc-2.25.orig/nptl/lowlevellock.c glibc-2.25/nptl/lowlevellock.c
--- glibc-2.25.orig/nptl/lowlevellock.c	2017-09-11 17:15:06.360004095 +0000
+++ glibc-2.25/nptl/lowlevellock.c	2017-09-11 22:18:57.889850357 +0000
@@ -26,7 +26,7 @@
 void
 __lll_lock_wait_private (int *futex)
 {
-  if (*futex == 2)
+  if (atomic_load(*futex) == 2)
     lll_futex_wait (futex, 2, LLL_PRIVATE); /* Wait if *futex == 2.  */
 
   while (atomic_exchange_acq (futex, 2) != 0)
@@ -39,7 +39,7 @@
 void
 __lll_lock_wait (int *futex, int private)
 {
-  if (*futex == 2)
+  if (atomic_load(*futex) == 2)
     lll_futex_wait (futex, 2, private); /* Wait if *futex == 2.  */
 
   while (atomic_exchange_acq (futex, 2) != 0)
diff -N -r --unified glibc-2.25.orig/nptl/pthread_cond_common.c glibc-2.25/nptl/pthread_cond_common.c
--- glibc-2.25.orig/nptl/pthread_cond_common.c	2017-09-11 17:15:06.389006169 +0000
+++ glibc-2.25/nptl/pthread_cond_common.c	2017-09-11 22:18:57.892850572 +0000
@@ -54,7 +54,7 @@
 __condvar_add_g1_start_relaxed (pthread_cond_t *cond, unsigned int val)
 {
   atomic_store_relaxed (&cond->__data.__g1_start,
-      atomic_load_relaxed (&cond->__data.__g1_start) + val);
+      orig_atomic_load_relaxed (&cond->__data.__g1_start) + val);
 }
 
 #else
diff -N -r --unified glibc-2.25.orig/nptl/pthread_join.c glibc-2.25/nptl/pthread_join.c
--- glibc-2.25.orig/nptl/pthread_join.c	2017-09-11 17:15:06.410007671 +0000
+++ glibc-2.25/nptl/pthread_join.c	2017-09-11 22:52:21.661783860 +0000
@@ -100,7 +100,7 @@
   if (__glibc_likely (result == 0))
     {
       /* We mark the thread as terminated and as joined.  */
-      pd->tid = -1;
+		atomic_store(pd->tid, -1);
 
       /* Store the return value if the caller is interested.  */
       if (thread_return != NULL)
diff -N -r --unified glibc-2.25.orig/nptl/pthread_mutex_setprioceiling.c glibc-2.25/nptl/pthread_mutex_setprioceiling.c
--- glibc-2.25.orig/nptl/pthread_mutex_setprioceiling.c	2017-09-11 17:15:06.412007815 +0000
+++ glibc-2.25/nptl/pthread_mutex_setprioceiling.c	2017-09-11 22:18:57.894850716 +0000
@@ -109,9 +109,9 @@
 
   int newlock = 0;
   if (locked)
-    newlock = (mutex->__data.__lock & ~PTHREAD_MUTEX_PRIO_CEILING_MASK);
-  mutex->__data.__lock = newlock
-			 | (prioceiling << PTHREAD_MUTEX_PRIO_CEILING_SHIFT);
+	  newlock = (atomic_load(mutex->__data.__lock) & ~PTHREAD_MUTEX_PRIO_CEILING_MASK);
+  atomic_store(mutex->__data.__lock, newlock
+			   | (prioceiling << PTHREAD_MUTEX_PRIO_CEILING_SHIFT));
   atomic_full_barrier ();
 
   lll_futex_wake (&mutex->__data.__lock, INT_MAX,
diff -N -r --unified glibc-2.25.orig/nptl/pthread_spin_lock.c glibc-2.25/nptl/pthread_spin_lock.c
--- glibc-2.25.orig/nptl/pthread_spin_lock.c	2017-09-11 17:15:06.442009961 +0000
+++ glibc-2.25/nptl/pthread_spin_lock.c	2017-09-11 22:18:57.897850932 +0000
@@ -54,12 +54,12 @@
 	{
 	  int wait = SPIN_LOCK_READS_BETWEEN_CMPXCHG;
 
-	  while (*lock != 0 && wait > 0)
+	  while (atomic_load(*lock) != 0 && wait > 0)
 	    --wait;
 	}
       else
 	{
-	  while (*lock != 0)
+		while (atomic_load(*lock) != 0)
 	    ;
 	}
     }
diff -N -r --unified glibc-2.25.orig/nptl/pthread_spin_unlock.c glibc-2.25/nptl/pthread_spin_unlock.c
--- glibc-2.25.orig/nptl/pthread_spin_unlock.c	2017-09-11 17:15:06.442009961 +0000
+++ glibc-2.25/nptl/pthread_spin_unlock.c	2017-09-11 22:18:57.899851075 +0000
@@ -24,6 +24,6 @@
 pthread_spin_unlock (pthread_spinlock_t *lock)
 {
   atomic_full_barrier ();
-  *lock = 0;
+  atomic_store(*lock, 0);
   return 0;
 }
diff -N -r --unified glibc-2.25.orig/nptl/register-atfork.c glibc-2.25/nptl/register-atfork.c
--- glibc-2.25.orig/nptl/register-atfork.c	2017-09-11 17:15:06.443010032 +0000
+++ glibc-2.25/nptl/register-atfork.c	2017-09-11 22:18:57.902851291 +0000
@@ -51,18 +51,19 @@
     {
       /* Search for an empty entry.  */
       for (i = 0; i < NHANDLER; ++i)
-	if (runp->mem[i].refcntr == 0)
+        if (atomic_load(runp->mem[i].refcntr) == 0)
 	  goto found;
     }
-  while ((runp = runp->next) != NULL);
+  while ((runp = atomic_load(runp->next)) != NULL);
 
   /* We have to allocate a new entry.  */
   runp = (struct fork_handler_pool *) calloc (1, sizeof (*runp));
   if (runp != NULL)
     {
       /* Enqueue the new memory pool into the list.  */
-      runp->next = fork_handler_pool.next;
-      fork_handler_pool.next = runp;
+		void* tmp = atomic_load(fork_handler_pool.next);
+      atomic_store(runp->next, tmp);
+      atomic_store(fork_handler_pool.next, runp);
 
       /* We use the last entry on the page.  This means when we start
 	 searching from the front the next time we will find the first
@@ -71,7 +72,7 @@
 
     found:
       result = &runp->mem[i];
-      result->refcntr = 1;
+      atomic_store(result->refcntr, 1);
       result->need_signal = 0;
     }
 
@@ -112,7 +113,7 @@
 __linkin_atfork (struct fork_handler *newp)
 {
   do
-    newp->next = __fork_handlers;
+	  newp->next = atomic_load(__fork_handlers);
   while (catomic_compare_and_exchange_bool_acq (&__fork_handlers,
 						newp, newp->next) != 0);
 }
@@ -124,10 +125,10 @@
   lll_lock (__fork_lock, LLL_PRIVATE);
 
   /* No more fork handlers.  */
-  __fork_handlers = NULL;
+  atomic_store(__fork_handlers, NULL);
 
   /* Free eventually allocated memory blocks for the object pool.  */
-  struct fork_handler_pool *runp = fork_handler_pool.next;
+  struct fork_handler_pool *runp = atomic_load(fork_handler_pool.next);
 
   memset (&fork_handler_pool, '\0', sizeof (fork_handler_pool));
 
@@ -138,7 +139,7 @@
   while (runp != NULL)
     {
       struct fork_handler_pool *oldp = runp;
-      runp = runp->next;
+      runp = atomic_load(runp->next);
       free (oldp);
     }
 }
diff -N -r --unified glibc-2.25.orig/nptl/unregister-atfork.c glibc-2.25/nptl/unregister-atfork.c
--- glibc-2.25.orig/nptl/unregister-atfork.c	2017-09-11 17:15:06.618022551 +0000
+++ glibc-2.25/nptl/unregister-atfork.c	2017-09-11 22:18:57.904851435 +0000
@@ -33,7 +33,7 @@
      We do not worry about other threads adding entries for this DSO
      right this moment.  If this happens this is a race and we can do
      whatever we please.  The program will crash anyway seen.  */
-  struct fork_handler *runp = __fork_handlers;
+  struct fork_handler *runp = atomic_load(__fork_handlers);
   struct fork_handler *lastp = NULL;
 
   while (runp != NULL)
@@ -42,7 +42,7 @@
     else
       {
 	lastp = runp;
-	runp = runp->next;
+	runp = atomic_load(runp->next);
       }
 
   if (runp == NULL)
@@ -77,12 +77,15 @@
 							 runp->next, runp)
 		  != 0)
 		{
-		  runp = __fork_handlers;
+          runp = atomic_load(__fork_handlers);
 		  goto again;
 		}
 	    }
 	  else
-	    lastp->next = runp->next;
+	  {
+		  void* tmp = atomic_load(runp->next);
+        atomic_store(lastp->next, tmp);
+	  }
 
 	  /* We cannot overwrite the ->next element now.  Put the deleted
 	     entries in a separate list.  */
@@ -94,7 +97,7 @@
       else
 	lastp = runp;
 
-      runp = runp->next;
+      runp = atomic_load(runp->next);
     }
   while (runp != NULL);
 
@@ -113,9 +116,9 @@
 	 wait for the last user.  */
       atomic_decrement (&deleted->handler->refcntr);
       unsigned int val;
-      while ((val = deleted->handler->refcntr) != 0)
+      while ((val = atomic_load(deleted->handler->refcntr)) != 0)
 	futex_wait_simple (&deleted->handler->refcntr, val, FUTEX_PRIVATE);
 
-      deleted = deleted->next;
+      deleted = atomic_load(deleted->next);
     }
 }
diff -N -r --unified glibc-2.25.orig/resolv/res_libc.c glibc-2.25/resolv/res_libc.c
--- glibc-2.25.orig/resolv/res_libc.c	2017-09-11 17:15:07.654096665 +0000
+++ glibc-2.25/resolv/res_libc.c	2017-09-11 22:18:57.907851650 +0000
@@ -28,7 +28,7 @@
 
 extern unsigned long long int __res_initstamp attribute_hidden;
 /* We have atomic increment operations on 64-bit platforms.  */
-#if __WORDSIZE == 64
+#if 0
 # define atomicinclock(lock) (void) 0
 # define atomicincunlock(lock) (void) 0
 # define atomicinc(var) catomic_increment (&(var))
diff -N -r --unified glibc-2.25.orig/stdlib/cxa_atexit.c glibc-2.25/stdlib/cxa_atexit.c
--- glibc-2.25.orig/stdlib/cxa_atexit.c	2017-09-11 17:15:08.086127569 +0000
+++ glibc-2.25/stdlib/cxa_atexit.c	2017-09-11 22:18:57.909851794 +0000
@@ -44,7 +44,7 @@
   new->func.cxa.arg = arg;
   new->func.cxa.dso_handle = d;
   atomic_write_barrier ();
-  new->flavor = ef_cxa;
+  atomic_store(new->flavor, ef_cxa);
   return 0;
 }
 
@@ -81,7 +81,7 @@
   for (l = *listp; l != NULL; p = l, l = l->next)
     {
       for (i = l->idx; i > 0; --i)
-	if (l->fns[i - 1].flavor != ef_free)
+		  if (atomic_load(l->fns[i - 1].flavor) != ef_free)
 	  break;
 
       if (i > 0)
@@ -123,7 +123,7 @@
   /* Mark entry as used, but we don't know the flavor now.  */
   if (r != NULL)
     {
-      r->flavor = ef_us;
+		atomic_store(r->flavor, ef_us);
       ++__new_exitfn_called;
     }
 
diff -N -r --unified glibc-2.25.orig/stdlib/cxa_finalize.c glibc-2.25/stdlib/cxa_finalize.c
--- glibc-2.25.orig/stdlib/cxa_finalize.c	2017-09-11 17:15:08.086127569 +0000
+++ glibc-2.25/stdlib/cxa_finalize.c	2017-09-11 22:18:57.911851938 +0000
@@ -70,7 +70,7 @@
 
       for (f = &funcs->fns[funcs->idx - 1]; f >= &funcs->fns[0]; --f)
 	if (d == NULL || d == f->func.cxa.dso_handle)
-	  f->flavor = ef_free;
+		atomic_store(f->flavor, ef_free);
     }
 
   /* Remove the registered fork handlers.  We do not have to
diff -N -r --unified glibc-2.25.orig/stdlib/exit.c glibc-2.25/stdlib/exit.c
--- glibc-2.25.orig/stdlib/exit.c	2017-09-11 17:15:08.098128427 +0000
+++ glibc-2.25/stdlib/exit.c	2017-09-11 22:18:57.914852154 +0000
@@ -52,7 +52,7 @@
 	{
 	  const struct exit_function *const f =
 	    &cur->fns[--cur->idx];
-	  switch (f->flavor)
+	  switch (atomic_load(f->flavor))
 	    {
 	      void (*atfct) (void);
 	      void (*onfct) (int status, void *arg);
diff -N -r --unified glibc-2.25.orig/stdlib/on_exit.c glibc-2.25/stdlib/on_exit.c
--- glibc-2.25.orig/stdlib/on_exit.c	2017-09-11 17:15:08.151132219 +0000
+++ glibc-2.25/stdlib/on_exit.c	2017-09-11 22:18:57.916852297 +0000
@@ -35,7 +35,7 @@
   new->func.on.fn = func;
   new->func.on.arg = arg;
   atomic_write_barrier ();
-  new->flavor = ef_on;
+  atomic_store(new->flavor, ef_on);
   return 0;
 }
 weak_alias (__on_exit, on_exit)
diff -N -r --unified glibc-2.25.orig/sysdeps/nptl/fork.c glibc-2.25/sysdeps/nptl/fork.c
--- glibc-2.25.orig/sysdeps/nptl/fork.c	2017-09-11 17:15:12.163419229 +0000
+++ glibc-2.25/sysdeps/nptl/fork.c	2017-09-11 22:18:57.919852513 +0000
@@ -63,12 +63,12 @@
   /* Run all the registered preparation handlers.  In reverse order.
      While doing this we build up a list of all the entries.  */
   struct fork_handler *runp;
-  while ((runp = __fork_handlers) != NULL)
+  while ((runp = atomic_load(__fork_handlers)) != NULL)
     {
       /* Make sure we read from the current RUNP pointer.  */
       atomic_full_barrier ();
 
-      unsigned int oldval = runp->refcntr;
+      unsigned int oldval = atomic_load(runp->refcntr);
 
       if (oldval == 0)
 	/* This means some other thread removed the list just after
@@ -103,7 +103,7 @@
 	  allp = newp;
 
 	  /* Advance to the next handler.  */
-	  runp = runp->next;
+	  runp = atomic_load(runp->next);
 	  if (runp == NULL)
 	    break;
 
@@ -214,7 +214,7 @@
 	     may have been bumped up by other threads doing a fork.
 	     We reset it to 1, to avoid waiting for non-existing
 	     thread(s) to release the count.  */
-	  allp->handler->refcntr = 1;
+	  atomic_store(allp->handler->refcntr, 1);
 
 	  /* XXX We could at this point look through the object pool
 	     and mark all objects not on the __fork_handlers list as
@@ -222,7 +222,7 @@
 	     while another thread called dlclose() and that call had
 	     to create a new list.  */
 
-	  allp = allp->next;
+	  allp = atomic_load(allp->next);
 	}
 
       /* Initialize the fork lock.  */
@@ -252,7 +252,7 @@
 	      && allp->handler->need_signal)
 	    futex_wake (&allp->handler->refcntr, 1, FUTEX_PRIVATE);
 
-	  allp = allp->next;
+	  allp = atomic_load(allp->next);
 	}
     }
 
diff -N -r --unified glibc-2.25.orig/sysdeps/nptl/lowlevellock.h glibc-2.25/sysdeps/nptl/lowlevellock.h
--- glibc-2.25.orig/sysdeps/nptl/lowlevellock.h	2017-09-11 17:15:12.166419443 +0000
+++ glibc-2.25/sysdeps/nptl/lowlevellock.h	2017-09-12 07:32:50.864131363 +0000
@@ -184,8 +184,8 @@
 #define lll_wait_tid(tid) \
   do {					\
     __typeof (tid) __tid;		\
-    while ((__tid = (tid)) != 0)	\
-      lll_futex_wait (&(tid), __tid, LLL_SHARED);\
+    while ((__tid = atomic_load(tid)) != 0 || mvee_should_sync_tid())	\
+		lll_futex_syscall(4, &(tid), __lll_private_flag(mvee_should_sync_tid() ? MVEE_FUTEX_WAIT_TID : FUTEX_WAIT, LLL_SHARED), __tid, NULL); \
   } while (0)
 
 extern int __lll_timedwait_tid (int *, const struct timespec *)
@@ -196,7 +196,7 @@
 #define lll_timedwait_tid(tid, abstime) \
   ({							\
     int __res = 0;					\
-    if ((tid) != 0)					\
+    if (atomic_load(tid) != 0)					\
       __res = __lll_timedwait_tid (&(tid), (abstime));	\
     __res;						\
   })
diff -N -r --unified glibc-2.25.orig/sysdeps/unix/sysv/linux/check_pf.c glibc-2.25/sysdeps/unix/sysv/linux/check_pf.c
--- glibc-2.25.orig/sysdeps/unix/sysv/linux/check_pf.c	2017-09-11 17:15:13.642525033 +0000
+++ glibc-2.25/sysdeps/unix/sysv/linux/check_pf.c	2017-09-11 22:18:57.922852729 +0000
@@ -269,7 +269,7 @@
   if (seen_ipv6 && result != NULL)
     {
       result->timestamp = get_nl_timestamp ();
-      result->usecnt = 2;
+      atomic_store(result->usecnt, 2);
       result->seen_ipv4 = seen_ipv4;
       result->seen_ipv6 = true;
       result->in6ailen = result_len;
@@ -348,7 +348,7 @@
       *in6ailen = data->in6ailen;
       *in6ai = data->in6ai;
 
-      if (olddata != NULL && olddata->usecnt > 0
+      if (olddata != NULL && atomic_load(olddata->usecnt) > 0
 	  && atomic_add_zero (&olddata->usecnt, -1))
 	free (olddata);
 
@@ -381,7 +381,7 @@
 	{
 	  __libc_lock_lock (lock);
 
-	  if (data->usecnt == 0)
+	  if (atomic_load(data->usecnt) == 0)
 	    /* Still unused.  */
 	    free (data);
 
diff -N -r --unified glibc-2.25.orig/sysdeps/unix/sysv/linux/lowlevellock-futex.h glibc-2.25/sysdeps/unix/sysv/linux/lowlevellock-futex.h
--- glibc-2.25.orig/sysdeps/unix/sysv/linux/lowlevellock-futex.h	2017-09-11 17:15:13.937546137 +0000
+++ glibc-2.25/sysdeps/unix/sysv/linux/lowlevellock-futex.h	2017-09-11 22:18:57.925852944 +0000
@@ -115,7 +115,7 @@
 		     nr_wake, nr_move, mutex, val)
 
 /* Returns non-zero if error happened, zero if success.  */
-#define lll_futex_wake_unlock(futexp, nr_wake, nr_wake2, futexp2, private) \
+#define orig_lll_futex_wake_unlock(futexp, nr_wake, nr_wake2, futexp2, private) \
   lll_futex_syscall (6, futexp,                                         \
 		     __lll_private_flag (FUTEX_WAKE_OP, private),       \
 		     nr_wake, nr_wake2, futexp2,                        \
