diff -r --unified linux-3.16.0.orig/arch/x86/kernel/ptrace.c linux-3.16.0/arch/x86/kernel/ptrace.c
--- linux-3.16.0.orig/arch/x86/kernel/ptrace.c	2015-02-26 16:39:57.498586829 -0800
+++ linux-3.16.0/arch/x86/kernel/ptrace.c	2015-02-26 16:40:33.414585585 -0800
@@ -1524,3 +1524,32 @@
 
 	user_enter();
 }
+
+/* TODO: move this to subarch so we can get this shit working on i386 */
+unsigned long arch_ptrace_get_program_counter()
+{
+  return getreg(current, 8 * 16); /* RIP = 16 */
+}
+
+/* This gets called at the syscall entry site */
+static unsigned char tmp_utcb_mask[] = {0xc0, 0x00, 0x78, 0x80, 0x09, 0x8f, 0x19, 0x00, 0x00, 0x01, 0x00, 0x00, 0xea, 0x9a, 0x12, 0xcc, 0x00, 0x09, 0x78, 0x00, 0x00, 0x00, 0x00, 0x21, 0xf8, 0x08, 0x00, 0x00, 0xcc, 0x00, 0x00, 0x08, 0x00, 0x00, 0x00, 0x00, 0x01, 0x90, 0x14, 0x00};
+unsigned char arch_ptrace_utcb_bypass(struct pt_regs * regs)
+{
+  /* check if the UTCB wants to bypass the kernel */
+  if (regs->r12 == 0xb00b135)
+    {
+      /* check the syscall no */
+      if (utcb_is_unchecked_syscall(current->mm->utcb_info.utcb_masked_syscalls, regs->orig_ax))
+	{
+	  /* final check: did this syscall originate from the registered UTCB? */
+	  unsigned long pc = arch_ptrace_get_program_counter();
+	  if (pc >= current->mm->utcb_info.utcb_start && pc <= current->mm->utcb_info.utcb_end)
+	    return 1;
+	}
+      
+      /* uh oh, the UTCB is being a bad boy. Refuse the bypass and clear r11 */
+      putreg(current, 8 * 3, 0x0); /* R12 = 3 */
+    }
+
+  return 0;
+}
diff -r --unified linux-3.16.0.orig/include/linux/mm_types.h linux-3.16.0/include/linux/mm_types.h
--- linux-3.16.0.orig/include/linux/mm_types.h	2015-02-26 16:40:14.490586241 -0800
+++ linux-3.16.0/include/linux/mm_types.h	2015-02-26 16:40:33.442585584 -0800
@@ -15,6 +15,7 @@
 #include <linux/page-flags-layout.h>
 #include <asm/page.h>
 #include <asm/mmu.h>
+#include <asm/unistd.h>
 
 #ifndef AT_VECTOR_SIZE_ARCH
 #define AT_VECTOR_SIZE_ARCH 0
@@ -343,6 +344,70 @@
 	atomic_long_t count[NR_MM_COUNTERS];
 };
 
+#ifndef ROUND_UP
+  #define ROUND_UP(x, multiple) ( (((long)(x)) + multiple - 1) & (~(multiple - 1)) )
+#endif
+
+#ifndef __NR_syscalls
+  /* this is the highest known syscall no for x86_64 in this kernel... */
+  #define __NR_syscalls __NR_seccomp
+#endif
+
+/* We're assuming that we're little endian here... */
+static inline unsigned char utcb_is_unchecked_syscall(unsigned char* mask, unsigned long syscall_no)
+{
+  unsigned long no_to_byte, bit_in_byte;
+
+  /* This is not very concise but the compiler will optimize it anyway... */
+  if (syscall_no > ROUND_UP(__NR_syscalls, 8))
+    return 0;
+
+  no_to_byte  = syscall_no / 8;
+  bit_in_byte = syscall_no % 8;
+
+  if (mask[no_to_byte] & (1 << (7 - bit_in_byte)))
+    return 1;
+  return 0;
+}
+
+static inline void utcb_set_unchecked_syscall(unsigned char* mask, unsigned long syscall_no, unsigned char unchecked)
+{
+  unsigned long no_to_byte, bit_in_byte;
+
+  if (syscall_no > ROUND_UP(__NR_syscalls, 8))
+    return;
+
+  no_to_byte  = syscall_no / 8;
+  bit_in_byte = syscall_no % 8;
+
+  if (unchecked)
+    mask[no_to_byte] |= (1 << (7 - bit_in_byte));
+  else
+    mask[no_to_byte] &= ~(1 << (7 - bit_in_byte));
+}
+
+/* MVEE UTCB patch */
+struct utcb_info {
+  /* This is a bitmask in which every bit represents exactly one syscall. 
+     If the bit is set to 1, then the corresponding syscall may bypass the ptracer under two conditions:
+
+     1) The syscall must have originated from the UTCB
+     2) The UTCB must pass the UTCB key as the 7th syscall argument
+  */
+  unsigned char utcb_masked_syscalls[ROUND_UP(__NR_syscalls, 8) / 8];
+
+  /* This is the randomly generated UTCB key that the UTCB can pass as the 7th syscall argument
+     to indicate that it wants the syscall to bypass the ptracer 
+  */
+  unsigned int  utcb_key;
+
+  /* These are the bounds of the UTCB executable section at the time of the UTCB
+     registration. The ptracer is responsible for ensuring that these bounds stay valid.
+  */
+  unsigned long utcb_start;
+  unsigned long utcb_end;
+};
+
 struct kioctx_table;
 struct mm_struct {
 	struct vm_area_struct *mmap;		/* list of VMAs */
@@ -456,6 +521,7 @@
 	bool tlb_flush_pending;
 #endif
 	struct uprobes_state uprobes_state;
+        struct utcb_info utcb_info;
 };
 
 static inline void mm_init_cpumask(struct mm_struct *mm)
diff -r --unified linux-3.16.0.orig/include/linux/ptrace.h linux-3.16.0/include/linux/ptrace.h
--- linux-3.16.0.orig/include/linux/ptrace.h	2015-02-26 16:40:14.386586244 -0800
+++ linux-3.16.0/include/linux/ptrace.h	2015-02-26 16:40:33.450585584 -0800
@@ -41,6 +41,10 @@
 #define PT_BLOCKSTEP_BIT	30
 #define PT_BLOCKSTEP		(1<<PT_BLOCKSTEP_BIT)
 
+/* MVEE UTCB patch */
+extern unsigned long arch_ptrace_get_program_counter(void);
+extern unsigned char arch_ptrace_utcb_bypass(struct pt_regs* regs);
+
 extern long arch_ptrace(struct task_struct *child, long request,
 			unsigned long addr, unsigned long data);
 extern int ptrace_readdata(struct task_struct *tsk, unsigned long src, char __user *dst, int len);
diff -r --unified linux-3.16.0.orig/include/linux/sched.h linux-3.16.0/include/linux/sched.h
--- linux-3.16.0.orig/include/linux/sched.h	2015-02-26 16:40:14.530586239 -0800
+++ linux-3.16.0/include/linux/sched.h	2015-02-26 16:40:33.450585584 -0800
@@ -1310,6 +1310,7 @@
 	/* Revert to default priority/policy when forking */
 	unsigned sched_reset_on_fork:1;
 	unsigned sched_contributes_to_load:1;
+        unsigned ptrace_ignored_current:1;
 
 	unsigned long atomic_flags; /* Flags needing atomic access. */
 
diff -r --unified linux-3.16.0.orig/include/linux/tracehook.h linux-3.16.0/include/linux/tracehook.h
--- linux-3.16.0.orig/include/linux/tracehook.h	2015-02-26 16:40:14.434586243 -0800
+++ linux-3.16.0/include/linux/tracehook.h	2015-02-26 16:40:33.454585584 -0800
@@ -62,6 +62,19 @@
 	if (!(ptrace & PT_PTRACED))
 		return 0;
 
+	/* this might be set at the syscall exit */
+	if (current->ptrace_ignored_current)
+	  {
+	    current->ptrace_ignored_current = 0;
+	    return 0;
+	  }
+	else if ((current->mm->utcb_info.utcb_key)
+	    && arch_ptrace_utcb_bypass(regs))
+	  {
+	    current->ptrace_ignored_current = 1;
+	    return 0;
+	  }
+
 	ptrace_notify(SIGTRAP | ((ptrace & PT_TRACESYSGOOD) ? 0x80 : 0));
 
 	/*
diff -r --unified linux-3.16.0.orig/include/uapi/linux/prctl.h linux-3.16.0/include/uapi/linux/prctl.h
--- linux-3.16.0.orig/include/uapi/linux/prctl.h	2015-02-26 16:40:16.154586183 -0800
+++ linux-3.16.0/include/uapi/linux/prctl.h	2015-02-26 16:40:33.462585584 -0800
@@ -152,4 +152,7 @@
 #define PR_SET_THP_DISABLE	41
 #define PR_GET_THP_DISABLE	42
 
+/* MVEE UTCB patch */
+#define PR_REGISTER_UTCB        0xb00b135
+
 #endif /* _LINUX_PRCTL_H */
diff -r --unified linux-3.16.0.orig/kernel/sys.c linux-3.16.0/kernel/sys.c
--- linux-3.16.0.orig/kernel/sys.c	2015-02-26 16:40:16.326586177 -0800
+++ linux-3.16.0/kernel/sys.c	2015-02-27 16:25:09.507897224 -0800
@@ -57,6 +57,7 @@
 /* Move somewhere else to avoid recompiling? */
 #include <generated/utsrelease.h>
 
+
 #include <asm/uaccess.h>
 #include <asm/io.h>
 #include <asm/unistd.h>
@@ -1821,6 +1822,175 @@
 	return error;
 }
 
+static unsigned long utcb_allowed_syscalls [] =
+  {
+    __NR_madvise,
+    __NR_getxattr,
+    __NR_lgetxattr,
+    __NR_fgetxattr,
+    __NR_listxattr,
+    __NR_llistxattr,
+    __NR_flistxattr,
+    __NR_getcwd,
+    __NR_ioprio_get,
+    __NR_read,
+    __NR_write,
+    __NR_readv,
+    __NR_writev,
+    __NR_pread64,
+    __NR_pwrite64,
+    __NR_preadv,
+    __NR_pwritev,
+    __NR_sendfile,
+    __NR_capget,
+    __NR_getitimer,
+    __NR_timer_gettime,
+    __NR_timer_getoverrun,
+    __NR_clock_gettime,
+    __NR_clock_getres,
+    __NR_sched_getscheduler,
+    __NR_sched_getparam,
+    __NR_sched_getaffinity,
+    __NR_sched_yield,
+    __NR_sched_get_priority_max,
+    __NR_sched_get_priority_min,
+    __NR_sched_rr_get_interval,
+    __NR_getpriority,
+    __NR_getresuid,
+    __NR_getresgid,
+    __NR_times,
+    __NR_getpgid,
+    __NR_getsid,
+    __NR_getgroups,
+    __NR_getrlimit,
+    __NR_getrusage,
+    __NR_getcpu,
+    __NR_gettimeofday,
+    __NR_getpid,
+    __NR_getppid,
+    __NR_getuid,
+    __NR_geteuid,
+    __NR_getgid,
+    __NR_getegid,
+    __NR_gettid,
+    __NR_getsockname,
+    __NR_getpeername,
+    __NR_sendto,
+    __NR_recvfrom,
+    __NR_getsockopt,
+    __NR_sendmsg,
+    __NR_recvmsg,
+    __NR_recvmmsg,
+    __NR_sendmmsg,
+    __NR_poll
+  };
+
+static int prctl_register_utcb(unsigned char __user* syscall_mask, 
+			       unsigned long syscall_mask_size)
+{
+  int error = 0, i, j;
+  struct vm_area_struct* vma = NULL;
+  unsigned long kernel_syscall_mask_size = ROUND_UP(__NR_syscalls, 8) / 8;
+  unsigned char syscall_mask_copy[kernel_syscall_mask_size];
+  unsigned char syscall_mask_check[kernel_syscall_mask_size];
+  struct mm_struct *mm = current->mm;
+  unsigned long program_counter;
+  unsigned int syscall_no;
+
+  /* this only makes sense if we're being ptraced... */
+  if (!current->ptrace)
+    {
+      printk("process %s (%d) attempted to register a UTCB but is not being ptraced\n", current->comm, current->pid);
+      return -ESRCH;
+    }
+
+  /* see if we already have a UTCB - this is racy 
+     need to recheck after we get the semaphore */
+  if (mm->utcb_info.utcb_key)
+    {
+      printk("process %s (%d) attempted to register a UTCB but it has registered a UTCB before\n", current->comm, current->pid);
+      return mm->utcb_info.utcb_key;
+    }
+
+  /* The UTCB may not mask calls that this kernel doesn't know */
+  if (syscall_mask_size > kernel_syscall_mask_size)
+    {
+      printk("process %s (%d) attempted to register a UTCB but the syscall_mask_size is invalid (%lu)\n", current->comm, current->pid, syscall_mask_size);
+      return -EINVAL;
+    }
+
+  /* OK. We did not have an initialized UTCB yet and the mask size seems valid. */
+  if (copy_from_user(syscall_mask_copy, syscall_mask, syscall_mask_size))
+    {
+      printk("process %s (%d) attempted to register a UTCB but the syscall_mask_address is invalid (%lx)\n", current->comm, current->pid, (unsigned long)syscall_mask);
+      return -EFAULT;
+    }
+
+  if (syscall_mask_size < kernel_syscall_mask_size)
+    memset(syscall_mask_copy + syscall_mask_size, 0, kernel_syscall_mask_size - syscall_mask_size);
+
+  /* Now check if the UTCB is trying to be naughty */
+  memcpy(syscall_mask_check, syscall_mask_copy, kernel_syscall_mask_size);
+
+  /* mask all syscalls that may actually bypass the ptracer */
+  for (i = 0; i < sizeof(utcb_allowed_syscalls) / sizeof(unsigned long); ++i)
+    utcb_set_unchecked_syscall(syscall_mask_check, utcb_allowed_syscalls[i], 0);
+
+  /* If the mask is non-null then the program is trying to do something nasty :) */
+  for (i = 0; i < kernel_syscall_mask_size; ++i)
+    {
+      if (syscall_mask_check[i])
+	{
+	  for (j = 0; j < 8; ++j)
+	    {
+	      if ((syscall_mask_check[i] >> j) & 0x1)
+		{
+		  syscall_no = i * 8 + (7 - j);
+		  break;
+		}
+	    }
+	  printk("process %s (%d) attempted to register a UTCB but the syscall_mask is invalid (syscall %d may not bypass the ptracer)\n", current->comm, current->pid, syscall_no);
+	  return -EPERM;
+	}
+    }
+
+  /* ok so the UTCB has not been initialized yet and the program is not trying to
+     unblock any system calls that we will not allow */
+  down_read(&mm->mmap_sem);
+
+  /* check once again if the UTCB has not been initialized yet. */
+  if (mm->utcb_info.utcb_key)
+    {
+      printk("process %s (%d) attempted to register a UTCB but it has just been registered\n", current->comm, current->pid);
+      goto out;
+    }
+
+  /* fetch the program counter and map it to the VMA */
+  program_counter = arch_ptrace_get_program_counter();
+
+  vma = find_vma(mm, program_counter);
+  if (!vma)
+    {
+      error = -EINVAL;
+      printk("process %s (%d) attempted to register a UTCB but we could not find the VMA associated with PC: %lx\n", current->comm, current->pid, program_counter);
+      goto out;
+    }
+  
+  /* OK! We have everything... */
+  mm->utcb_info.utcb_start = vma->vm_start;
+  mm->utcb_info.utcb_end = vma->vm_end;
+  memcpy(mm->utcb_info.utcb_masked_syscalls, syscall_mask_copy, kernel_syscall_mask_size);
+  /* TODO: Implement me! This should generate a random 32-bit integer between 1 and 0x7FFFFFFF */
+  mm->utcb_info.utcb_key = 0xb00b135;
+  error = mm->utcb_info.utcb_key;
+
+  printk("process %s (%d) registered a UTCB at: %lx-%lx with key %x\n", current->comm, current->pid, vma->vm_start, vma->vm_end, mm->utcb_info.utcb_key);
+
+ out:
+  up_read(&mm->mmap_sem);
+  return error;
+}
+
 #ifdef CONFIG_CHECKPOINT_RESTORE
 static int prctl_get_tid_address(struct task_struct *me, int __user **tid_addr)
 {
@@ -2011,6 +2180,10 @@
 			me->mm->def_flags &= ~VM_NOHUGEPAGE;
 		up_write(&me->mm->mmap_sem);
 		break;
+	/* MVEE UTCB patch */
+	case PR_REGISTER_UTCB:
+                error = prctl_register_utcb((unsigned char __user*)arg2, arg3);
+                break;
 	default:
 		error = -EINVAL;
 		break;
